

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>CoSpar basics with synthetic data &mdash; CoSpar 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="About scVelo" href="about.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About scVelo</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">CoSpar basics with synthetic data</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>CoSpar basics with synthetic data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar_notebooks/blob/master/20201121-Bifurcation_model_staticBC-CoSapr_basic.ipynb" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="CoSpar-basics-with-synthetic-data">
<h1>CoSpar basics with synthetic data<a class="headerlink" href="#CoSpar-basics-with-synthetic-data" title="Permalink to this headline">¶</a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">new_path</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span><span class="o">+</span><span class="s2">&quot;/CoSpar/&quot;</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_path</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>

<span class="c1"># Define the data path as well as the figure path. Will be used to load or save data/figures.</span>
<span class="c1"># If they do not exist yet, create them.</span>
<span class="n">data_path</span><span class="o">=</span><span class="s1">&#39;bifur_StaticBC_data&#39;</span> <span class="c1"># A relative path to save data.</span>
<span class="n">figure_path</span><span class="o">=</span><span class="s1">&#39;bifur_StaticBC_figure&#39;</span> <span class="c1"># A relative path to save figures.</span>
<span class="o">!</span>mkdir -p <span class="nv">$data_path</span>
<span class="o">!</span>mkdir -p <span class="nv">$figure_path</span>
</pre></div>
</div>
</div>
<div class="section" id="Initialization-and-Preprocessing">
<h2>Initialization and Preprocessing<a class="headerlink" href="#Initialization-and-Preprocessing" title="Permalink to this headline">¶</a></h2>
<p>In order for this to work, it is mandatory that you provide</p>
<ul class="simple">
<li><p><strong>RNA_count_matrix</strong>: np.array or sparse matrix, shape (n_cell, n_gene)</p></li>
<li><p><strong>gene_names</strong>: list of gene names, shape (n_genes,)</p></li>
<li><p><strong>time_info</strong>: np.array of time annotation in string, shape (n_cell,)</p></li>
</ul>
<p>The clonal data is optional. If not provided, cospar generates the transition map based on state information only. A clone means a barcode that labels cells. We allow a cell have multiple barcodes, thus belonging to multiple clones.</p>
<ul class="simple">
<li><p><strong>X_clone</strong>: np.array or sparse matrix, shape (n_cell, n_clone)</p></li>
</ul>
<p>It is optional that you also provide the following data. However, they can be generated in further preprocessing after initializing the anadata object.</p>
<ul class="simple">
<li><p><strong>X_pca</strong>: pca matrixs, np.array, shape (n_cell, n_pcs)</p></li>
<li><p><strong>X_emb</strong>: two-dimensional embedding, np.array, shape (n_cell, 2)</p></li>
<li><p><strong>state_info</strong>: state annotation for each cell, np.array, shape (n_cell, 1)</p></li>
</ul>
<div class="section" id="Loading-data">
<h3>Loading data<a class="headerlink" href="#Loading-data" title="Permalink to this headline">¶</a></h3>
<p>Download raw data from this link: * <a class="reference external" href="https://kleintools.hms.harvard.edu/tools/downloads/cospar/bifur_adata_preprocessed.h5ad">https://kleintools.hms.harvard.edu/tools/downloads/cospar/bifur_adata_preprocessed.h5ad</a></p>
<p>and put it to the folder of data_path</p>
<p>Or use wget:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">!</span>wget https://kleintools.hms.harvard.edu/tools/downloads/cospar/bifur_adata_preprocessed.h5ad
<span class="o">!</span>mv bifur_adata_preprocessed.h5ad <span class="nv">$data_path</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig_0</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{data_path}</span><span class="s1">/bifur_adata_preprocessed.h5ad&#39;</span><span class="p">)</span>

<span class="c1"># This is just a name to indicate this data.</span>
<span class="c1"># Will be used to name the saved results. Can be arbitrary.</span>
<span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;bifur&#39;</span>

<span class="n">RNA_count_matrix</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">X</span> <span class="c1"># np.array or sparse matrix, shape (n_cell, n_gene)</span>
<span class="n">gene_names</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">var_names</span> <span class="c1"># List of gene names, shape (n_genes,)</span>

<span class="c1"># Clonal data matrix, np.array or sparse matrix, shape: (n_cell, n_clone)</span>
<span class="n">X_clone</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_clone&#39;</span><span class="p">]</span>

<span class="c1"># 2-d embedding, np.array, shape: (n_cell, 2)</span>
<span class="n">X_emb</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span>

<span class="c1"># A vector of cluster id for each cell, np.array, shape: (n_cell, 2),</span>
<span class="n">state_info</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;state_info&#39;</span><span class="p">]</span>

<span class="c1"># principle component matrix, np.array, shape: (n_cell, n_pcs)</span>
<span class="n">X_pca</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">]</span>

<span class="c1"># A vector of time info, np.array of string, shape: (n_cell,)</span>
<span class="n">time_info</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;time_info&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Initialize-the-adata-structure">
<h3>Initialize the adata structure<a class="headerlink" href="#Initialize-the-adata-structure" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">initialize_adata_object</span><span class="p">(</span><span class="n">RNA_count_matrix</span><span class="p">,</span><span class="n">gene_names</span><span class="p">,</span><span class="n">time_info</span><span class="p">,</span>
     <span class="n">X_clone</span><span class="o">=</span><span class="n">X_clone</span><span class="p">,</span>
        <span class="n">X_pca</span><span class="o">=</span><span class="n">X_pca</span><span class="p">,</span><span class="n">X_emb</span><span class="o">=</span><span class="n">X_emb</span><span class="p">,</span><span class="n">state_info</span><span class="o">=</span><span class="n">state_info</span><span class="p">,</span>
        <span class="n">data_path</span><span class="o">=</span><span class="n">data_path</span><span class="p">,</span><span class="n">figure_path</span><span class="o">=</span><span class="n">figure_path</span><span class="p">,</span><span class="n">data_des</span><span class="o">=</span><span class="n">data_des</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[Errno 17] File exists: &#39;bifur_StaticBC_data&#39;
[Errno 17] File exists: &#39;bifur_StaticBC_figure&#39;
All time points: {&#39;1&#39;, &#39;2&#39;}
Time points with clonal info: {&#39;1&#39;, &#39;2&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2474 × 50
    obs: &#39;time_info&#39;, &#39;state_info&#39;
    uns: &#39;data_des&#39;, &#39;data_path&#39;, &#39;figure_path&#39;, &#39;clonal_time_points&#39;
    obsm: &#39;X_clone&#39;, &#39;X_pca&#39;, &#39;X_umap&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Preprocessing-and-dimension-reduction">
<h3>Preprocessing and dimension reduction<a class="headerlink" href="#Preprocessing-and-dimension-reduction" title="Permalink to this headline">¶</a></h3>
<p>You could just use scanpy to do all the preprocessing and dimension reduction.</p>
<p>We assume the data quality control and preprocessing is performed, say by using scanpy. Such a preprocessing could include: 1) remove low-quality cells/genes; 2) count normalization; 3) remove doublet cells; 4) regress out cell cycle effects.</p>
<p>We provide functions for excluding cell-cycle correlated genes among highly variable genes, and for generating PCA components, dimension reduction based on UMAP, and clustering based on leiden.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_12_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_12_0.png" />
</div>
</div>
</div>
<div class="section" id="Save-preprocessed-data-before-Transition-map-inference.">
<h3>Save preprocessed data before Transition map inference.<a class="headerlink" href="#Save-preprocessed-data-before-Transition-map-inference." title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">save_data</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Save pre-computed&quot;</span><span class="p">)</span>
    <span class="n">data_des</span><span class="o">=</span><span class="n">adata_orig</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;data_des&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1">#adata.uns[&#39;state_trajectory&#39;]={} # need to set to empty, otherwise, it does not work</span>
    <span class="c1">#adata.uns[&#39;fate_trajectory&#39;]={} # need to set to empty, otherwise, it does not work</span>
    <span class="n">adata_orig</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{data_path}</span><span class="s1">/</span><span class="si">{data_des}</span><span class="s1">_adata_preprocessed.h5ad&#39;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">data_des</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Raw-clonal-data-analysis-(without-using-state-information)">
<h3>Raw clonal data analysis (without using state information)<a class="headerlink" href="#Raw-clonal-data-analysis-(without-using-state-information)" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_16_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_16_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;2&#39;</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[],</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_17_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_17_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;2&#39;</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barcode_heatmap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[],</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_18_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_18_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">select_fate_cluster</span><span class="o">=</span><span class="s1">&#39;1&#39;</span>
<span class="n">clonal_fate_bias</span><span class="p">,</span><span class="n">clone_id</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonal_fate_bias</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">select_fate_cluster</span><span class="p">,</span><span class="n">N_resampling</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current clone id: 0
Current clone id: 5
Current clone id: 10
Current clone id: 15
Current clone id: 20
Current clone id: 25
Current clone id: 30
Current clone id: 35
Current clone id: 40
Current clone id: 45
Current clone id: 50
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_19_1.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_19_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_19_2.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_19_2.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Transition-map-inference">
<h2>Transition map inference<a class="headerlink" href="#Transition-map-inference" title="Permalink to this headline">¶</a></h2>
<p>An adata with only sub-sampled cell states at given time points will be returned, with Tmap appended at adata.uns[‘transition_map’]</p>
<p>The first to run it, it will generate the Similarity matrices and save them. It could be slow (around half hours) for data with around 10,000 cells. Once the similarity matrices are generated, it takes a few minutes to generate the Tmap for 10000 cells. The runtime of the algorithm currently does not scale up well at this moment.</p>
<div class="section" id="If-clones-have-multiple-time-points">
<h3>If clones have multiple time points<a class="headerlink" href="#If-clones-have-multiple-time-points" title="Permalink to this headline">¶</a></h3>
<p><em>Purpose:</em> Compute transition map for re-sampled clonal data with both state and lineage information. We assume that the lineage information spans at least two time points.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;All time points: {set(adata_orig.obs[&#39;time_info&#39;])}&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Time points with clonal info: {set(adata_orig.uns[&#39;clonal_time_points&#39;])}&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All time points: {&#39;1&#39;, &#39;2&#39;}
Time points with clonal info: {&#39;1&#39;, &#39;2&#39;}
</pre></div></div>
</div>
<p>Generate the transition map</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">time</span>

<span class="n">t</span><span class="o">=</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">False</span>
<span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1">#</span>
<span class="n">selected_clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span><span class="s1">&#39;2&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clonal_time_points</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span>
            <span class="n">CoSpar_KNN</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">noise_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span>
             <span class="n">use_full_Smatrix</span><span class="o">=</span><span class="n">use_full_Smatrix</span><span class="p">,</span><span class="n">use_all_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Used--time: {time.time()-t}&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------Step 1: Select time points---------
Clonal cell fraction (day 1-2): 1.0
Clonal cell fraction (day 2-1): 1.0
Numer of cells that are clonally related -- day 1: 854  and day 2: 1620
Valid clone number &#39;FOR&#39; post selection 52
Cell number=2474, Clone number=52
-------Step 2: Compute the full Similarity matrix if necessary---------
-------Step 3: Optimize the transition map recursively---------
---------Compute the transition map-----------
Compute similarity matrix: load existing data
Time elapsed:  0.018784761428833008
Time elapsed:  0.054964303970336914
Time elapsed:  0.01058816909790039
Time elapsed:  0.040583133697509766
Compute similarity matrix: load existing data
Time elapsed:  0.017374038696289062
Time elapsed:  0.03009200096130371
Time elapsed:  0.014258861541748047
Time elapsed:  0.04777717590332031
Compute similarity matrix: load existing data
Time elapsed:  0.010631799697875977
Time elapsed:  0.02845597267150879
Time elapsed:  0.01053476333618164
Time elapsed:  0.03804421424865723
Current iteration: 0
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.03924679756164551
Phase II: time elapsed --  0.05923199653625488
Current iteration: 1
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.022300004959106445
Phase II: time elapsed --  0.033738136291503906
Current iteration: 2
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.021209001541137695
Phase II: time elapsed --  0.03739500045776367
No need for Final Smooth
----Demultiplexed transition map----
Clone normalization
Time point pair index: 0
Clone id: 0
Used--time: 1.6666450500488281
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_26_1.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_26_1.png" />
</div>
</div>
<p>Generate demultiplexed map within each clone (Optional, as this map has been generated already)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">run_demultiplex</span><span class="o">=</span><span class="kc">True</span>
<span class="k">if</span> <span class="n">run_demultiplex</span><span class="p">:</span>
    <span class="n">demulti_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1"># This threshold should be smaller, ass the map has been further smoothed to expand to more states.</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_intraclone_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">demulti_threshold</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Clone normalization
Time point pair index: 0
Clone id: 0
</pre></div></div>
</div>
</div>
<div class="section" id="If-clones-are-from-a-single-time-point">
<h3>If clones are from a single time point<a class="headerlink" href="#If-clones-are-from-a-single-time-point" title="Permalink to this headline">¶</a></h3>
<p><em>Purpose</em>: Infer transition map from scRNAseq data where cells at one time point are clonally labeled. After initializing the map by either <em>OT</em> method or <em>HighVar</em> method, We jointly infer the likely clonal ancestors and the transition map between cell states in these two time points.</p>
<p>Parameters relevant for cell state selection: <em>initial_time_points, clonal_time_point, use_full_Smatrix</em>.</p>
<p>Choose the initialization method, and set the corresponding parameters.</p>
<ul class="simple">
<li><p>‘HighVar’: is faster and robust to batch effect. Related parameters: <em>HighVar_gene_pctl</em>.</p></li>
<li><p>‘OT’: tend to be more accurate, but very slow and not reliable under batch effect. Key parameters: <em>OT_epsilon</em>.</p></li>
</ul>
<p>Parameters relevant for CoSpar itself: <em>smooth_array, normalization_mode, CoSpar_KNN, noise_threshold, Clone_update_iter_N</em>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;All time points: {set(adata_orig.obs[&#39;time_info&#39;])}&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Time points with clonal info: {set(adata_orig.uns[&#39;clonal_time_points&#39;])}&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All time points: {&#39;1&#39;, &#39;2&#39;}
Time points with clonal info: {&#39;1&#39;, &#39;2&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span>
<span class="n">clonal_time_point</span><span class="o">=</span><span class="s1">&#39;2&#39;</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_one_time_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">clonal_time_point</span><span class="p">,</span>
     <span class="n">Clone_update_iter_N</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">initialize_method</span><span class="o">=</span><span class="s1">&#39;OT&#39;</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span>
                        <span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">compute_new</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 1
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Load pre-computed shortest path distance matrix
Load pre-computed custon OT matrix
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap
original clone shape: (2474, 52)
After excluding zero-sized clones: (2474, 52)
Sort clones by size (small to large)
Inferring early clonal states: current clone id 0
Inferring early clonal states: current clone id 50
Warning: early break. Final clone id: 52
---------Compute the transition map-----------
Compute similarity matrix: load existing data
Time elapsed:  0.01004481315612793
Time elapsed:  0.029703378677368164
Time elapsed:  0.011106014251708984
Time elapsed:  0.029217958450317383
Compute similarity matrix: load existing data
Time elapsed:  0.009356021881103516
Time elapsed:  0.026854991912841797
Time elapsed:  0.012832164764404297
Time elapsed:  0.0292971134185791
Compute similarity matrix: load existing data
Time elapsed:  0.010351896286010742
Time elapsed:  0.03115391731262207
Time elapsed:  0.014170169830322266
Time elapsed:  0.030233144760131836
Current iteration: 0
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.0360870361328125
Phase II: time elapsed --  0.057615041732788086
Current iteration: 1
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.015066146850585938
Phase II: time elapsed --  0.031071901321411133
Current iteration: 2
Use SM=10
Clone normalization
Time point pair index: 0
Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.013494014739990234
Phase II: time elapsed --  0.02470707893371582
No need for Final Smooth
----Demultiplexed transition map----
Clone normalization
Time point pair index: 0
Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 1.728844165802002
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Bridge-steps">
<h2>Bridge steps<a class="headerlink" href="#Bridge-steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Adjusting-state_annotation-(optional)">
<h3>Adjusting state_annotation (optional)<a class="headerlink" href="#Adjusting-state_annotation-(optional)" title="Permalink to this headline">¶</a></h3>
<p>We combine marker gene expression and time information to select cell states for a particular fate</p>
<p>The changes are not confirmed until you set <em>confirm_change=True</em></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># #&#39;NPC2&#39;,&#39;SFTPC&#39;,&#39;CLDN18&#39;,&#39;CEBPD&#39;,&#39;NAPSA&#39;,&#39;PRSS1&#39;,&#39;PGC&#39;,&#39;CXCL2&#39;</span>
<span class="c1"># confirm_change=False</span>
<span class="c1"># marker_genes=[&#39;NPC2&#39;,&#39;SFTPC&#39;]</span>
<span class="c1"># cs.pp.refine_state_info_by_marker_genes(adata,marker_genes,gene_threshold=0.2,</span>
<span class="c1">#     selected_time_points=[&#39;D27&#39;],new_cluster_name=&#39;new&#39;,add_neighbor_N=10,confirm_change=confirm_change)</span>
</pre></div>
</div>
</div>
<p>Use leiden clustering for states at selected time points to generate cluster ID for cell states</p>
<p>The changes are not confirmed until you set <em>confirm_change=True</em></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">False</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">refine_state_info_by_leiden_clustering</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">],</span><span class="n">leiden_resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                                    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="n">confirm_change</span><span class="p">,</span><span class="n">cluster_name_prefix</span><span class="o">=</span><span class="s1">&#39;T&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_39_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_39_0.png" />
</div>
</div>
</div>
<div class="section" id="Save-or-load-pre-computed-maps">
<h3>Save or load pre-computed maps<a class="headerlink" href="#Save-or-load-pre-computed-maps" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">save_data</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">save_map</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">load_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">load_data</span><span class="p">:</span>

    <span class="c1">#data_des=&#39;bifur_TTC_Timet*1*2&#39;</span>
    <span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;bifur_TwoTimeClone_t*1*2&#39;</span>
    <span class="c1">#data_des=&#39;bifur_OneTimeClone_t*1*2&#39;</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Load data: data_des=&#39;</span><span class="si">{data_des}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;</span><span class="si">{data_path}</span><span class="s1">/</span><span class="si">{data_des}</span><span class="s1">_adata_with_transition_map.h5ad&#39;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 2474 × 50
    obs: &#39;state_info&#39;, &#39;time_info&#39;
    uns: &#39;data_path&#39;, &#39;data_des&#39;, &#39;figure_path&#39;, &#39;Tmap_cell_id_t1&#39;, &#39;Tmap_cell_id_t2&#39;, &#39;clonal_cell_id_t1&#39;, &#39;clonal_cell_id_t2&#39;, &#39;sp_idx&#39;, &#39;transition_map&#39;, &#39;OT_transition_map&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_clone&#39;
</pre></div></div>
</div>
<p>Directly access the maps etc.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">export_variables</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">export_variables</span><span class="p">:</span>
    <span class="n">transition_map</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;transition_map&#39;</span><span class="p">]</span>
    <span class="c1">#demultiplexed_map=adata.uns[&#39;demultiplexed_map&#39;]</span>
    <span class="n">state_annote_0</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;state_info&#39;</span><span class="p">]</span>
    <span class="n">Tmap_cell_id_t1</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;Tmap_cell_id_t1&#39;</span><span class="p">]</span> <span class="c1"># initial state indices for the transition map</span>
    <span class="n">Tmap_cell_id_t2</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;Tmap_cell_id_t2&#39;</span><span class="p">]</span> <span class="c1"># later state indices for the transition map</span>
    <span class="n">x_emb</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">y_emb</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">data_des</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;data_des&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">time_info</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;time_info&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Plotting">
<h2>Plotting<a class="headerlink" href="#Plotting" title="Permalink to this headline">¶</a></h2>
<p>There are some common parameters</p>
<ul class="simple">
<li><p><strong>used_map_name</strong>: <em>{‘transition_map’, ‘demultiplexed_map’, ‘OT_transition_map’, ‘HighVar_transition_map’}</em></p></li>
<li><p><strong>selected_fates</strong>: selected fate clusters for things like fate map construction, DGE analysis etc. The <em>selected_fates</em> allows nested structure, e.g., <em>selected_fates=[[‘0’,‘2’],‘1’]</em> means combined cluster ‘0-2’ and cluster ‘1’</p></li>
<li><p><strong>map_backwards</strong>: <em>True</em>, assuming that the <em>selected_fates</em> are cell states belonging to later time points (t2), and map for the initial cell states; <em>False</em>, map from early states to later states.</p></li>
<li><p><strong>plot_time_points</strong>: select states at given time points for the analysis</p></li>
</ul>
<p>For more information, use <em>‘? function_name’</em> to read the documentation.</p>
<div class="section" id="Transition-profiles-for-single-cells">
<h3>Transition profiles for single cells<a class="headerlink" href="#Transition-profiles-for-single-cells" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition_probability</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_46_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_46_0.png" />
</div>
</div>
</div>
<div class="section" id="Fate-map">
<h3>Fate map<a class="headerlink" href="#Fate-map" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_transition_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: [&#39;transition_map&#39;, &#39;OT_transition_map&#39;]
Availabel clusters: [&#39;0&#39;, &#39;1&#39;, &#39;-1&#39;]
Availabel time points: [&#39;1&#39;, &#39;2&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span><span class="s1">&#39;0&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">plot_fate_bias</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_49_0.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_49_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_49_1.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_49_1.png" />
</div>
</div>
</div>
<div class="section" id="Binary-fate-choice">
<h3>Binary fate choice<a class="headerlink" href="#Binary-fate-choice" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">binary_fate_choice</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span><span class="s1">&#39;1&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">include_target_states</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(854, 2)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_51_1.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_51_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_51_2.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_51_2.png" />
</div>
</div>
</div>
</div>
<div class="section" id="Apply-the-naive-method">
<h2>Apply the naive method<a class="headerlink" href="#Apply-the-naive-method" title="Permalink to this headline">¶</a></h2>
<p>Assume uniform transition within the same clone</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_naive_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">binary_fate_choice</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span><span class="s1">&#39;1&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;naive_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">include_target_states</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(854, 2)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_53_1.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_53_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_53_2.png" src="_images/20201121-Bifurcation_model_staticBC-CoSapr_basic_53_2.png" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="about.html" class="btn btn-neutral float-left" title="About scVelo" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>