

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>CoSpar basics &mdash; CoSpar 0.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Hematopoietic dataset" href="20210121_all_hematopoietic_data.html" />
    <link rel="prev" title="Getting Started" href="getting_started.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="about.html">About CoSpar</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">CoSpar basics</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="20210121_all_hematopoietic_data.html">Hematopoietic dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_merge_tags.html">Reprogramming dataset (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_no_merge_tags.html">Reprogramming dataset (dynamic barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_lung_data.html">Lung dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_static_barcoding.html">Synthetic data (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_dynamic_barcoding.html">Synthetic data (dynamic barcoding)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>CoSpar basics</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar_notebooks/blob/master/20210121_cospar_tutorial.ipynb" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="CoSpar-basics">
<h1>CoSpar basics<a class="headerlink" href="#CoSpar-basics" title="Permalink to this headline">¶</a></h1>
<p>Below, we walk you through typical steps of analysis. We use the subsampled hematopoiesis dataset so that this tutorial an be run within a few minutes. For further demonstration, please see notebooks at the <strong>Examples</strong> section.</p>
<p><strong>Contents</strong>:</p>
<ul class="simple">
<li><p>Initialization</p></li>
<li><p>Preprocessing and dimension reduction (optional)</p></li>
<li><p>Raw clonal data analysis (without using state information)</p></li>
<li><p>Transition map inference</p></li>
<li><p>Plotting</p></li>
<li><p>Miscellaneous</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">new_path</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span><span class="o">+</span><span class="s2">&quot;/CoSpar/&quot;</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_path</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">3</span> <span class="c1">#range: 0 (error),1 (warning),2 (info),3 (hint). Set it at a given value will print information at and below its level.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span> <span class="c1"># figure setting.</span>
</pre></div>
</div>
</div>
<p>Each dataset should have its own folder to avoid conflict.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># set the directory for figures and data. If not existed yet, it will be created automtaically.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">data_path</span><span class="o">=</span><span class="s1">&#39;data_cospar&#39;</span> <span class="c1">#data_path</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">figure_path</span><span class="o">=</span><span class="s1">&#39;fig_path&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="Initialization">
<h2>Initialization<a class="headerlink" href="#Initialization" title="Permalink to this headline">¶</a></h2>
<p>The first step is to construct an adata object that stores all relevant data. The input data are:</p>
<ul class="simple">
<li><p><strong>RNA_count_matrix</strong>: shape (n_cell, n_gene), mandatory, stored at <code class="docutils literal notranslate"><span class="pre">adata.X</span></code></p></li>
<li><p><strong>gene_names</strong>: list of gene names, shape (n_genes,), mandatory, stored at <code class="docutils literal notranslate"><span class="pre">adata.var_names</span></code></p></li>
<li><p><strong>time_info</strong>: list of time annotation in string, shape (n_cell,), mandatory, stored at <code class="docutils literal notranslate"><span class="pre">adata.obs['time_info']</span></code></p></li>
<li><p><strong>X_clone</strong>: clonal label for each cell in the form of np.array or sparse matrix, shape (n_cell, n_clone), optional, stored at <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_clone']</span></code></p></li>
<li><p><strong>X_pca</strong>: PCA matrixs, shape (n_cell, n_pcs), optional, stored at <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_pca']</span></code></p></li>
<li><p><strong>X_emb</strong>: two-dimensional embedding, shape (n_cell, 2), optional, stored at <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_emb']</span></code></p></li>
<li><p><strong>state_info</strong>: state annotation for each cell, shape (n_cell, 1), optional, stored at <code class="docutils literal notranslate"><span class="pre">adata.obsm['state_info']</span></code></p></li>
</ul>
<p>The clonal data <code class="docutils literal notranslate"><span class="pre">X_clone</span></code> is optional. If not provided, cospar generates the transition map based on state information only. A clone means a barcode that labels cells. We allow a cell have multiple barcodes, thus belonging to multiple clones.</p>
<p><code class="docutils literal notranslate"><span class="pre">X_pca</span></code>, <code class="docutils literal notranslate"><span class="pre">X_emb</span></code>, and <code class="docutils literal notranslate"><span class="pre">state_info</span></code> are also optional. They can be generated in downstream preprocessing and dimension reduction steps.</p>
<p>We provide a function to help initializing this object. We mimic this step by loading a preprocessed dataset, extracting each component, and then re-assembling them using our initialization function.</p>
<p>Load an existing dataset (if you have pre-processed data, you can load it with <code class="docutils literal notranslate"><span class="pre">cs.hf.read(file_name)</span></code>)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig_0</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">hematopoiesis_subsampled</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Extract each component</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># This is just a name to indicate this data.</span>
<span class="c1"># Will be used to name the saved results. Can be arbitrary.</span>
<span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;blood&#39;</span>

<span class="n">RNA_count_matrix</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">X</span> <span class="c1"># np.array or sparse matrix, shape (n_cell, n_gene)</span>
<span class="n">gene_names</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">var_names</span> <span class="c1"># List of gene names, shape (n_genes,)</span>

<span class="c1"># Clonal data matrix, np.array or sparse matrix, shape: (n_cell, n_clone)</span>
<span class="n">X_clone</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_clone&#39;</span><span class="p">]</span>

<span class="c1"># 2-d embedding, np.array, shape: (n_cell, 2)</span>
<span class="n">X_emb</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_emb&#39;</span><span class="p">]</span>

<span class="c1"># A vector of cluster id for each cell, np.array, shape: (n_cell, 2),</span>
<span class="n">state_info</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;state_info&#39;</span><span class="p">]</span>

<span class="c1"># principle component matrix, np.array, shape: (n_cell, n_pcs)</span>
<span class="n">X_pca</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">]</span>

<span class="c1"># A vector of time info, np.array of string, shape: (n_cell,)</span>
<span class="n">time_info</span><span class="o">=</span><span class="n">adata_orig_0</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;time_info&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>Now, initialize the adata object</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">initialize_adata_object</span><span class="p">(</span><span class="n">RNA_count_matrix</span><span class="p">,</span><span class="n">gene_names</span><span class="p">,</span><span class="n">time_info</span><span class="p">,</span>
     <span class="n">X_clone</span><span class="o">=</span><span class="n">X_clone</span><span class="p">,</span><span class="n">X_pca</span><span class="o">=</span><span class="n">X_pca</span><span class="p">,</span><span class="n">X_emb</span><span class="o">=</span><span class="n">X_emb</span><span class="p">,</span><span class="n">state_info</span><span class="o">=</span><span class="n">state_info</span><span class="p">,</span><span class="n">data_des</span><span class="o">=</span><span class="n">data_des</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All time points: {&#39;4&#39;, &#39;6&#39;, &#39;2&#39;}
Time points with clonal info: {&#39;4&#39;, &#39;6&#39;, &#39;2&#39;}
</pre></div></div>
</div>
<p>This is the structure of the initialized adata object</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 7438 × 25289
    obs: &#39;time_info&#39;, &#39;state_info&#39;
    uns: &#39;data_des&#39;, &#39;clonal_time_points&#39;
    obsm: &#39;X_clone&#39;, &#39;X_pca&#39;, &#39;X_emb&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Preprocessing-and-dimension-reduction-(optional)">
<h2>Preprocessing and dimension reduction (optional)<a class="headerlink" href="#Preprocessing-and-dimension-reduction-(optional)" title="Permalink to this headline">¶</a></h2>
<p>If the data do not yet have one of <code class="docutils literal notranslate"><span class="pre">X_pca</span></code>, <code class="docutils literal notranslate"><span class="pre">X_emb</span></code>, or <code class="docutils literal notranslate"><span class="pre">state_info</span></code>, you will need to run the preprocessing and dimension reduction.</p>
<p>Select highly variable genes. Before this, there is count normalization.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_highly_variable_genes</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">normalized_counts_per_cell</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span><span class="n">min_counts</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
        <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_gene_vscore_pctl</span><span class="o">=</span><span class="mi">85</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding highly variable genes...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_18_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_18_1.png" style="width: 278px; height: 220px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Keeping 2489 genes
</pre></div></div>
</div>
<p>Compute for each gene its correlation with a set of cell cycle genes. There is a default cell cycle gene set for mouse. You might need to use your own genes. This step is optional, but recommended.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">remove_cell_cycle_correlated_genes</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
adata.var[&#39;highly_variable&#39;] not updated.
Please choose corr_threshold properly, and set confirm_change=True
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_20_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_20_1.png" style="width: 296px; height: 248px;" />
</div>
</div>
<p>Now, actually confirm the change at a specific cutoff threshold.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">remove_cell_cycle_correlated_genes</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">corr_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of selected non-cycling highly variable genes: 2481
Remove 8 cell cycle correlated genes.
adata.var[&#39;highly_variable&#39;] updated
</pre></div></div>
</div>
<p>Compute the <code class="docutils literal notranslate"><span class="pre">X_pca</span></code>, <code class="docutils literal notranslate"><span class="pre">X_emb</span></code>, and <code class="docutils literal notranslate"><span class="pre">state_info</span></code> (from clustering).</p>
<p><code class="docutils literal notranslate"><span class="pre">X_pca</span></code> will be used to build the similarity matrix later. <code class="docutils literal notranslate"><span class="pre">X_emb</span></code> is only used for visualization. You can also pass your favoriate embedding directly to <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_emb']</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">update</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">update</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_X_pca</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">n_pca_comp</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
    <span class="c1">#cs.pp.get_X_emb(adata_orig,n_neighbors=20,umap_min_dist=0.3) #Do not run this, as we want to keep the original embedding</span>
    <span class="c1">#cs.pp.get_state_info(adata_orig,leiden_resolution=0.5) # Do not run this, as we want to keep the original state annotation.</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_25_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_25_0.png" style="width: 470px; height: 274px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: []
Availabel clusters: [&#39;undiff&#39;, &#39;pDC&#39;, &#39;Erythroid&#39;, &#39;Neutrophil&#39;, &#39;Meg&#39;, &#39;Neu_Mon&#39;, &#39;Lymphoid&#39;, &#39;Ccr7_DC&#39;, &#39;Eos&#39;, &#39;Monocyte&#39;, &#39;Mast&#39;, &#39;Baso&#39;]
Availabel time points: [&#39;4&#39;, &#39;6&#39;, &#39;2&#39;]
Clonal time points: [&#39;4&#39;, &#39;6&#39;, &#39;2&#39;]
</pre></div></div>
</div>
<p>You can choose to save preprocessed data. It can be loaded using <code class="docutils literal notranslate"><span class="pre">cs.hf.read(file_name)</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">save</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">save_preprocessed_adata</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="Raw-clonal-data-analysis-(without-using-state-information)">
<h3>Raw clonal data analysis (without using state information)<a class="headerlink" href="#Raw-clonal-data-analysis-(without-using-state-information)" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span><span class="n">color_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">],</span><span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_30_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_30_0.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_30_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_30_1.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_30_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_30_2.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span> <span class="s1">&#39;Mast&#39;</span><span class="p">,</span> <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span> <span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span> <span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span> <span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;pDC&#39;</span><span class="p">,</span> <span class="s1">&#39;Eos&#39;</span><span class="p">,</span> <span class="s1">&#39;Meg&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_31_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_31_0.png" style="width: 358px; height: 283px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barcode_heatmap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_32_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_32_0.png" style="width: 284px; height: 433px;" />
</div>
</div>
<p>This step is slow and optional. The time is linearly proportional to the multiplication of clone number and N_resampling. It takes half a minute for the given parameters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">select_fate_cluster</span><span class="o">=</span><span class="s1">&#39;Monocyte&#39;</span>
<span class="n">clonal_fate_bias</span><span class="p">,</span><span class="n">clone_id</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonal_fate_bias</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">select_fate_cluster</span><span class="p">,</span>
            <span class="n">clone_size_thresh</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">N_resampling</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">compute_new</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_34_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_34_0.png" style="width: 283px; height: 246px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_34_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_34_1.png" style="width: 302px; height: 265px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Transition-map-inference">
<h2>Transition map inference<a class="headerlink" href="#Transition-map-inference" title="Permalink to this headline">¶</a></h2>
<p>As mentioned in <strong>Getting Started</strong>, there are several methods for map inference. Below, we illustrate the one that uses multiple clonal time points and integrate state and clonal information together.</p>
<p>The first time it runs, it will compute the similarity matrix at different smooth rounds and save them. Then, it will infer the transition map. It takes 2 mins to run for the first time, and 11 s for later runs.</p>
<p>Some key parameters: * <strong>smooth_array</strong>: a list of numbers defining the smooth rounds at each iteration of coherent sparsity optimization. Its length determins the number of iteration. We found that 3 rounds of iteration is generally sufficient. It is better to use a number at the multiple of 5, i.e., 5, 10, 15, 20,…</p>
<ul class="simple">
<li><p><strong>noise_threshold</strong>: noise threshold to remove spurious transitions in the updated transition map, in the range [0,1].</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1">#</span>
<span class="n">selected_clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clonal_time_points</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span>
        <span class="n">noise_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------Step 1: Select time points---------
--&gt; Clonal cell fraction (day 2-4): 1.0
--&gt; Clonal cell fraction (day 4-6): 1.0
--&gt; Clonal cell fraction (day 4-2): 0.3892466194462331
--&gt; Clonal cell fraction (day 6-4): 1.0
--&gt; Numer of cells that are clonally related -- day 2: 286  and day 4: 1209
--&gt; Numer of cells that are clonally related -- day 4: 3106  and day 6: 4046
Valid clone number &#39;FOR&#39; post selection 500
Cell number=7438, Clone number=500
-------Step 2: Compute the full Similarity matrix if necessary---------
-------Step 3: Optimize the transition map recursively---------
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.10666298866271973
--&gt; Time elapsed:  0.330460786819458
--&gt; Time elapsed:  0.09987902641296387
--&gt; Time elapsed:  0.3382446765899658
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.09616303443908691
--&gt; Time elapsed:  0.27065300941467285
--&gt; Time elapsed:  0.08978414535522461
--&gt; Time elapsed:  0.3071787357330322
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.06744384765625
--&gt; Time elapsed:  0.2183530330657959
--&gt; Time elapsed:  0.06598091125488281
--&gt; Time elapsed:  0.19800972938537598
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.5561578273773193
Phase II: time elapsed --  1.0665109157562256
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.45022106170654297
Phase II: time elapsed --  0.8973500728607178
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.45832180976867676
Phase II: time elapsed --  0.9774909019470215
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
-----------Total used time: 14.393146991729736 s ------------
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_38_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_38_1.png" style="width: 363px; height: 241px;" />
</div>
</div>
<p>We can see that after this step, we have two maps: <code class="docutils literal notranslate"><span class="pre">'transition_map'</span></code> and <code class="docutils literal notranslate"><span class="pre">'intraclone_transition_map'</span></code>. Both of them can be used for downstream analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: [&#39;transition_map&#39;, &#39;intraclone_transition_map&#39;]
Availabel clusters: [&#39;undiff&#39;, &#39;pDC&#39;, &#39;Erythroid&#39;, &#39;Neutrophil&#39;, &#39;Meg&#39;, &#39;Neu_Mon&#39;, &#39;Lymphoid&#39;, &#39;Ccr7_DC&#39;, &#39;Eos&#39;, &#39;Monocyte&#39;, &#39;Mast&#39;, &#39;Baso&#39;]
Availabel time points: [&#39;4&#39;, &#39;6&#39;, &#39;2&#39;]
Clonal time points: [&#39;4&#39;, &#39;6&#39;, &#39;2&#39;]
</pre></div></div>
</div>
</div>
<div class="section" id="Save-or-load-pre-computed-data">
<h2>Save or load pre-computed data<a class="headerlink" href="#Save-or-load-pre-computed-data" title="Permalink to this headline">¶</a></h2>
<p>This can be used to save adata with maps computed from different tools or parameters. Usually, different parameter choices will result in different ‘data_des’, a prefix to identify the anndata. Saving an adata would print the ‘data_des’, which can be used to load the corresponding adata.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">save_data</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">save_map</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">load_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">load_data</span><span class="p">:</span>
    <span class="c1"># updated Jan 27, 2021</span>
    <span class="c1">#data_des=&#39;blood_TwoTimeClone_t*4*6&#39;</span>
    <span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;blood_TwoTimeClone_t*2*4*6&#39;</span>
    <span class="c1">#data_des=&#39;LARRY_sp500_ranking1_TwoTimeClone_t*4*6&#39;</span>
    <span class="c1">#data_des=&#39;LARRY_sp500_ranking1_TwoTimeClone_t*2*4*6&#39;</span>
    <span class="c1">#data_des=&#39;LARRY_sp500_ranking1_OneTimeClone_t*4*6&#39;</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">load_saved_adata_with_key</span><span class="p">(</span><span class="n">data_des</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Plotting">
<h2>Plotting<a class="headerlink" href="#Plotting" title="Permalink to this headline">¶</a></h2>
<p>There are some common parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">used_map_name</span></code> (str). Determines which transition map to use for analysis. Choices: {‘transition_map’, ‘intraclone_transition_map’, ‘OT_transition_map’, ‘HighVar_transition_map’,’clonal_transition_map’}</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selected_fates</span></code> (list of str). Selected clusters to aggregate differentiation dynamics and visualize fate bias etc.. The selected_fates allows nested structure, e.g., selected_fates=[‘1’, [‘0’, ‘2’]] selects two clusters: cluster ‘1’ and the other that combines ‘0’ and ‘2’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">map_backwards</span></code> (bool). We can analyze either the forward transitions, i.e., where the selected states or clusters are going (map_backwards=False), or the backward transitions, i.e., where these selected states or clusters came from (map_backwards=False). The latter is more useful, and is the default.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">plot_time_points</span></code> (list). List of time points to use. By default, all are used.</p></li>
</ul>
<div class="section" id="Plotting-transition-profiles-for-single-cells">
<h3>Plotting transition profiles for single cells<a class="headerlink" href="#Plotting-transition-profiles-for-single-cells" title="Permalink to this headline">¶</a></h3>
<p>First, check the forward transitions (i.e., future states) from the <code class="docutils literal notranslate"><span class="pre">'transition_map'</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">]</span>

<span class="n">map_backwards</span><span class="o">=</span><span class="kc">False</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="n">map_backwards</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_46_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_46_0.png" style="width: 281px; height: 433px;" />
</div>
</div>
<p>Now, backward transitions (i.e., past states) from the same map</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">]</span>
<span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="n">map_backwards</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_48_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_48_0.png" style="width: 281px; height: 433px;" />
</div>
</div>
<p>Finally, switch to the <code class="docutils literal notranslate"><span class="pre">'intraclone_transition_map'</span></code>, and check backward transitions. We can see that the second id (5) does not have corresponding clonally related initial states</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">]</span>
<span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="n">map_backwards</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_50_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_50_0.png" style="width: 281px; height: 433px;" />
</div>
</div>
</div>
<div class="section" id="Fate-map">
<h3>Fate map<a class="headerlink" href="#Fate-map" title="Permalink to this headline">¶</a></h3>
<p>Inspect the backward transitions, and ask where the selected fate clusters are at the previous time point</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_53_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_53_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
<p>Now, do the same thing with <code class="docutils literal notranslate"><span class="pre">'intraclone_transition_map'</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_55_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_55_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
<p>As a comparison, if we use only the clonal information to construct the transition map, this is what is looks like:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_clonal_info_alone</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;weinreb&#39;</span><span class="p">)</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;clonal_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Used uni-potent clone fraction 0.472
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_57_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_57_1.png" style="width: 288px; height: 435px;" />
</div>
</div>
</div>
<div class="section" id="Relative-fate-bias">
<h3>Relative fate bias<a class="headerlink" href="#Relative-fate-bias" title="Permalink to this headline">¶</a></h3>
<div class="line-block">
<div class="line">First, check the intrinsic fate bias of initial states, which is basically the predicted fate map normalized by the expected fate probability. The null expectation is</div>
<div class="line">is the fraction of cells at later time point that are in the corresponding cluster. A state has an intrinsic fate bias for each fate cluster.</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_intrinsic</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_60_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_60_0.png" style="width: 528px; height: 230px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_60_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_60_1.png" style="width: 600px; height: 266px;" />
</div>
</div>
<p>Next, check the fate bias of initial states defined by competition between two fate clusters A and B, i.e., how strongly A is favored than B.</p>
<p>Only states with fate probabilities satisfying this criterion will be shown:</p>
<ul class="simple">
<li><p>P(A)+P(B)&gt;sum_fate_prob_thresh</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_62_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_62_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_62_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_62_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
</div>
<div class="section" id="Dynamic-trajectory-inference">
<h3>Dynamic trajectory inference<a class="headerlink" href="#Dynamic-trajectory-inference" title="Permalink to this headline">¶</a></h3>
<p>One way to define the dynamic trajectory is simply mapping a given fate cluster backward in time. The whole trajectory across multiple time points will be saved at <code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code>. This method performs best if there are multiple clonal time points.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_65_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_65_0.png" style="width: 733px; height: 237px;" />
</div>
</div>
<p>In the second approach, we use the intrinsic fate bias to define the dynamic trajectory and also the ancestor population. It requires input of two selected fates A and B, and states with bias for A higher than bias_threshold_A will be selected as ancestor states for fate A; so does for fate B. The selected cell states will be stored at <code class="docutils literal notranslate"><span class="pre">adata.uns['Cell_group_A']</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.uns['Cell_group_B']</span></code>. The selected states along with the corresponding fate cluster will be stored at
<code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code> for inspecting dynamic gene expression.</p>
<p>Setting <code class="docutils literal notranslate"><span class="pre">avoid_target_states=True</span></code> will remove ancestor states already inside the given fate cluster.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_intrinsic_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold_A</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">bias_threshold_B</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_67_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_67_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<p>We can also infer the dynamic trajectory and ancestor population using the fate bias from binary fate competition. Here, fate bias is a scalar between (0,1) at each state. Selected ancestor population satisfies:</p>
<ul class="simple">
<li><p>Prob(A)+Prob(B)&gt;sum_fate_prob_thresh;</p></li>
<li><p>Ancestor states for A: Bias&gt;0.5+bias_threshold</p></li>
<li><p>Ancestor states for B: bias&lt;0.5+bias_threshold</p></li>
</ul>
<p>They will be stored at <code class="docutils literal notranslate"><span class="pre">adata.uns['Cell_group_A']</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.uns['Cell_group_B']</span></code>. The selected states along with the corresponding fate cluster will be stored at <code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code> for inspecting dynamic gene expression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_competition_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_69_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_69_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
</div>
<div class="section" id="Differential-genes-for-two-ancestor-groups">
<h3>Differential genes for two ancestor groups<a class="headerlink" href="#Differential-genes-for-two-ancestor-groups" title="Permalink to this headline">¶</a></h3>
<p>It would be interesting to see what genes are differentially expressed between these two ancestor population, which might drive the fate bifurcation.</p>
<p>The two population can be accessed at <code class="docutils literal notranslate"><span class="pre">adata.obs['Cell_group_A']</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.obs['Cell_group_B']</span></code>. We provide a simple differentiation gene expression analysis that use Wilcoxon rank-sum test to calculate P values, followed by Benjamini-Hochberg correction. You can always use your own method.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># This step requires that adata.obs[&#39;Cell_group_A&#39;] and adata.obs[&#39;Cell_group_B&#39;] exist.</span>
<span class="n">dge_gene_A</span><span class="p">,</span> <span class="n">dge_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">differential_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">plot_gene_N</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_72_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_72_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_72_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_72_1.png" style="width: 709px; height: 171px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_72_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_72_2.png" style="width: 709px; height: 171px;" />
</div>
</div>
</div>
<div class="section" id="Gene-trend-along-the-dynamic-trajectory">
<h3>Gene trend along the dynamic trajectory<a class="headerlink" href="#Gene-trend-along-the-dynamic-trajectory" title="Permalink to this headline">¶</a></h3>
<p>Any one of the 3 dynamic trajectory inference methods will save the trajectory at <code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code>. We can calculate the pseudotime along this trajectory and plot the gene expression along this pseudo time.</p>
<p>This method requires that <code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code> exists, and the selected fate is in the pre-computed list.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gata1&#39;</span><span class="p">,</span><span class="s1">&#39;Mpo&#39;</span><span class="p">,</span> <span class="s1">&#39;Elane&#39;</span><span class="p">,</span> <span class="s1">&#39;S100a8&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compute_new</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fig_width</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">gene_exp_percentile</span><span class="o">=</span><span class="mi">99</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">plot_raw_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: fig_path/blood_TwoTimeClone_t*2*4*6_fate_trajectory_pseutoTime_gene_expression_Neutrophil_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_76_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_76_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_76_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_76_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
<p>The method outputs a new adata object with the selected cell states. This can be used to run your favoriate pseudotime analysis methods.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_selected</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Trying to set attribute `.uns` of view, copying.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_78_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_78_1.png" style="width: 470px; height: 274px;" />
</div>
</div>
</div>
<div class="section" id="Fate-coupling-of-the-transition-map">
<h3>Fate coupling of the transition map<a class="headerlink" href="#Fate-coupling-of-the-transition-map" title="Permalink to this headline">¶</a></h3>
<p>The inferred transition map can be used to estimate differentiation coupling between different fate clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fate_array</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">celltype_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mDC&#39;</span><span class="p">,</span>  <span class="s1">&#39;Mast&#39;</span><span class="p">,</span> <span class="s1">&#39;Meg&#39;</span><span class="p">,</span> <span class="s1">&#39;pDC&#39;</span><span class="p">,</span> <span class="s1">&#39;Eos&#39;</span><span class="p">,</span> <span class="s1">&#39;Ly&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery&#39;</span><span class="p">,</span> <span class="s1">&#39;Baso&#39;</span><span class="p">,</span> <span class="s1">&#39;Neu&#39;</span><span class="p">,</span> <span class="s1">&#39;Mon&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">fate_array</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
                        <span class="n">rename_selected_fates</span><span class="o">=</span><span class="n">celltype_names</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_81_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_81_0.png" style="width: 359px; height: 283px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Miscellaneous">
<h2>Miscellaneous<a class="headerlink" href="#Miscellaneous" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Transition-map-from-state-information-and-the-last-clonal-time-point">
<h3>Transition map from state information and the last clonal time point<a class="headerlink" href="#Transition-map-from-state-information-and-the-last-clonal-time-point" title="Permalink to this headline">¶</a></h3>
<p>After initializing the map by either <em>OT</em> method or <em>HighVar</em> method, We jointly infer the likely clonal ancestors and the transition map between cell states in across given time points.</p>
<p>You need to choose the initialization method, and set the corresponding parameters.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">OT</span></code>: optional transport based method. It tends to be more accurate than <code class="docutils literal notranslate"><span class="pre">HighVar</span></code>, but not reliable under batch differences between time points. Key parameters:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">OT_epsilon</span></code> for the entropic regularization</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">OT_cost</span></code>: {‘GED’, ‘SPD’}, method for computing the cost function. ‘GED’ uses simple gene expression distances. It is faster. ‘SPD’ uses the shortest path distances. It is slower, but more accurate.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">HighVar</span></code>: a method that converts highly variable genes into pseudo clones and run coherent sparsity optimization to generate initialized map. Although it is not as accurate as <code class="docutils literal notranslate"><span class="pre">OT</span></code>, it is robust to batch effect across time points, and is used to analyze the lung dataset.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">HighVar_gene_pctl</span></code>: percentile threshold to select highly variable genes. Range: [0,100]. A higher value selects more variable genes.</p></li>
</ul>
</li>
</ul>
<p>Let us try the OT method that uses gene expression distancs. It takes around 1 min to run.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">]</span> <span class="c1">#[&#39;2&#39;,&#39;4&#39;]</span>
<span class="n">clonal_time_point</span><span class="o">=</span><span class="s1">&#39;6&#39;</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_one_time_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">clonal_time_point</span><span class="p">,</span><span class="n">Clone_update_iter_N</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">initialize_method</span><span class="o">=</span><span class="s1">&#39;OT&#39;</span><span class="p">,</span><span class="n">OT_cost</span><span class="o">=</span><span class="s1">&#39;GED&#39;</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 4
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Finishing computing gene expression distance, used time 5.845627307891846
Load pre-computed custon OT matrix
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (7152, 500)
--&gt; After excluding zero-sized clones at t2: (7152, 500)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
--&gt; Inferring early clonal states: current clone id 400
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.07742595672607422
--&gt; Time elapsed:  0.11342000961303711
--&gt; Time elapsed:  0.07841730117797852
--&gt; Time elapsed:  0.1302661895751953
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.0654458999633789
--&gt; Time elapsed:  0.09871602058410645
--&gt; Time elapsed:  0.07186508178710938
--&gt; Time elapsed:  0.10639524459838867
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.05588507652282715
--&gt; Time elapsed:  0.08137392997741699
--&gt; Time elapsed:  0.08660697937011719
--&gt; Time elapsed:  0.08424496650695801
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.35403895378112793
Phase II: time elapsed --  0.7963449954986572
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.21891403198242188
Phase II: time elapsed --  0.633296012878418
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.2083740234375
Phase II: time elapsed --  0.5867331027984619
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 10.520820140838623
-----------Total used time: 17.815898895263672 s ------------
</pre></div></div>
</div>
<p>Now, check the result of the initialized map <code class="docutils literal notranslate"><span class="pre">OT_transition_map</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span>
    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span><span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_89_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_89_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_89_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_89_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
<p>After joint optimization, the final result of the transition map is much better:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span>
    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_91_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_91_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_91_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_91_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
</div>
<div class="section" id="Transition-map-from-state-information-alone">
<h3>Transition map from state information alone<a class="headerlink" href="#Transition-map-from-state-information-alone" title="Permalink to this headline">¶</a></h3>
<p>This is the same as <code class="docutils literal notranslate"><span class="pre">cs.tmap.infer_Tmap_from_one_time_clones</span></code>, except that we only compute the initialized map.</p>
<p>Let us try <code class="docutils literal notranslate"><span class="pre">HighVar</span></code> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">]</span> <span class="c1">#[&#39;2&#39;,&#39;4&#39;]</span>
<span class="n">target_time_point</span><span class="o">=</span><span class="s1">&#39;6&#39;</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_state_info_alone</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">target_time_point</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s1">&#39;HighVar&#39;</span><span class="p">,</span><span class="n">HighVar_gene_pctl</span><span class="o">=</span><span class="mi">85</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 4
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use highly variable genes to construct pseudo-clones, and apply CoSpar to generate initialized map!
HighVar-v0: avoid cells that have been selected
----------------
Step a: find the commonly shared highly variable genes
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_94_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_94_1.png" style="width: 278px; height: 220px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_94_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_94_2.png" style="width: 278px; height: 220px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Highly varable gene number at t1 is 2136, Highly varable gene number at t2 is 2245
Common gene set is 978
----------------
Step b: convert the shared highly variable genes into clonal info
No cells left for assignment, total used genes=850
----------------
Step c: compute the transition map based on clonal info from highly variable genes
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.1092829704284668
--&gt; Time elapsed:  0.14886212348937988
--&gt; Time elapsed:  0.07950091361999512
--&gt; Time elapsed:  0.11178112030029297
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.06927180290222168
--&gt; Time elapsed:  0.10271787643432617
--&gt; Time elapsed:  0.07293009757995605
--&gt; Time elapsed:  0.10834097862243652
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.07581591606140137
--&gt; Time elapsed:  0.08312821388244629
--&gt; Time elapsed:  0.0559239387512207
--&gt; Time elapsed:  0.10451817512512207
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.20914101600646973
Phase II: time elapsed --  0.6100800037384033
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.1894700527191162
Phase II: time elapsed --  0.5436420440673828
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.18822312355041504
Phase II: time elapsed --  0.5829341411590576
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from highly variable genes, used time 26.875142097473145
-----------Total used time: 27.308759212493896 s ------------
</pre></div></div>
</div>
<p>You can see that this method is not as accurate for this dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span>
    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;HighVar_transition_map&#39;</span><span class="p">,</span><span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_96_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_96_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_96_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_96_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
</div>
<div class="section" id="Gene-expression">
<h3>Gene expression<a class="headerlink" href="#Gene-expression" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_on_manifold</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_genes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gata2&#39;</span><span class="p">,</span><span class="s1">&#39;Mpo&#39;</span><span class="p">],</span><span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">savefig</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_98_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_98_0.png" style="width: 277px; height: 208px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_98_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_98_1.png" style="width: 277px; height: 208px;" />
</div>
</div>
</div>
<div class="section" id="Refine-state-annotation-by-marker-genes">
<h3>Refine state annotation by marker genes<a class="headerlink" href="#Refine-state-annotation-by-marker-genes" title="Permalink to this headline">¶</a></h3>
<p>The goal here is to refine <code class="docutils literal notranslate"><span class="pre">adata.obs['state_info']</span></code>. First, explore parameters to find satisfactory annotation.</p>
<p>In this method, a state is selected if it expresses all genes in the list of marker_genes, and the expression are above the relative threshold express_threshold. You can also specify which time point you want to focus on. In addition, we also include cell states neighboring to these valid states to smooth the selection (controlled by <code class="docutils literal notranslate"><span class="pre">add_neighbor_N</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">False</span>
<span class="n">marker_genes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Mpo&#39;</span><span class="p">,</span> <span class="s1">&#39;Elane&#39;</span><span class="p">,</span> <span class="s1">&#39;S100a8&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">refine_state_info_by_marker_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">marker_genes</span><span class="p">,</span>
    <span class="n">selected_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">new_cluster_name</span><span class="o">=</span><span class="s1">&#39;new&#39;</span><span class="p">,</span><span class="n">add_neighbor_N</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="n">confirm_change</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_101_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_101_0.png" style="width: 247px; height: 202px;" />
</div>
</div>
<p>Now, confirm changes to <code class="docutils literal notranslate"><span class="pre">adata.obs['state_info']</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">True</span>
<span class="n">marker_genes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Mpo&#39;</span><span class="p">,</span> <span class="s1">&#39;Elane&#39;</span><span class="p">,</span> <span class="s1">&#39;S100a8&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">refine_state_info_by_marker_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">marker_genes</span><span class="p">,</span>
    <span class="n">selected_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">new_cluster_name</span><span class="o">=</span><span class="s1">&#39;new&#39;</span><span class="p">,</span><span class="n">add_neighbor_N</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="n">confirm_change</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Change state annotation at adata.obs[&#39;state_info&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_103_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_103_1.png" style="width: 247px; height: 202px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_103_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_103_2.png" style="width: 470px; height: 274px;" />
</div>
</div>
</div>
<div class="section" id="Refine-state-annotation-by-clustering-states-at-given-time-points">
<h3>Refine state annotation by clustering states at given time points<a class="headerlink" href="#Refine-state-annotation-by-clustering-states-at-given-time-points" title="Permalink to this headline">¶</a></h3>
<p>First, explore the parameters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">False</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">refine_state_info_by_leiden_clustering</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">],</span>
            <span class="n">leiden_resolution</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="n">confirm_change</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Error: Selected time points not available. Please select from [&#39;4&#39;, &#39;6&#39;]
</pre></div></div>
</div>
<p>Once you are happy with the result, confirm the change to <code class="docutils literal notranslate"><span class="pre">adata.obs['state_info']</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">True</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">refine_state_info_by_leiden_clustering</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">],</span>
                    <span class="n">leiden_resolution</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="n">confirm_change</span><span class="p">,</span><span class="n">cluster_name_prefix</span><span class="o">=</span><span class="s1">&#39;day2_&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Error: Selected time points not available. Please select from [&#39;4&#39;, &#39;6&#39;]
</pre></div></div>
</div>
</div>
<div class="section" id="Differential-gene-expression-between-two-clusters">
<h3>Differential gene expression between two clusters<a class="headerlink" href="#Differential-gene-expression-between-two-clusters" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">diff_gene_A</span><span class="p">,</span><span class="n">diff_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">differential_genes_for_given_fates</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
                                    <span class="n">gene_N</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">plot_gene_N</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_110_0.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_110_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_110_1.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_110_1.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_cospar_tutorial_110_2.png" class="no-scaled-link" src="_images/20210121_cospar_tutorial_110_2.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">diff_gene_A</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>pv</th>
      <th>mean_1</th>
      <th>mean_2</th>
      <th>ratio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16</th>
      <td>Ngp</td>
      <td>4.184225e-182</td>
      <td>1.684322</td>
      <td>324.910828</td>
      <td>-6.923776</td>
    </tr>
    <tr>
      <th>96</th>
      <td>Camp</td>
      <td>1.401590e-73</td>
      <td>0.132501</td>
      <td>24.727663</td>
      <td>-4.505735</td>
    </tr>
    <tr>
      <th>66</th>
      <td>Ltf</td>
      <td>5.324550e-86</td>
      <td>0.108736</td>
      <td>24.162815</td>
      <td>-4.504305</td>
    </tr>
    <tr>
      <th>12</th>
      <td>S100a9</td>
      <td>6.193967e-200</td>
      <td>33.714066</td>
      <td>442.656708</td>
      <td>-3.675852</td>
    </tr>
    <tr>
      <th>13</th>
      <td>S100a8</td>
      <td>2.892061e-198</td>
      <td>32.819645</td>
      <td>263.530884</td>
      <td>-2.967503</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>339</th>
      <td>Prss57</td>
      <td>2.689917e-24</td>
      <td>0.036761</td>
      <td>1.080616</td>
      <td>-1.004927</td>
    </tr>
    <tr>
      <th>314</th>
      <td>Ccdc125</td>
      <td>8.511859e-26</td>
      <td>0.562560</td>
      <td>2.135035</td>
      <td>-1.004570</td>
    </tr>
    <tr>
      <th>299</th>
      <td>Cnn2</td>
      <td>3.391266e-27</td>
      <td>0.903385</td>
      <td>2.791486</td>
      <td>-0.994196</td>
    </tr>
    <tr>
      <th>552</th>
      <td>Anxa1</td>
      <td>2.409673e-13</td>
      <td>7.341900</td>
      <td>15.615431</td>
      <td>-0.994076</td>
    </tr>
    <tr>
      <th>471</th>
      <td>Cybb</td>
      <td>4.089828e-16</td>
      <td>15.634762</td>
      <td>31.941101</td>
      <td>-0.985688</td>
    </tr>
  </tbody>
</table>
<p>100 rows × 5 columns</p>
</div></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="20210121_all_hematopoietic_data.html" class="btn btn-neutral float-right" title="Hematopoietic dataset" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="getting_started.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>