

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Hematopoietic dataset &mdash; CoSpar 0.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Reprogramming dataset (static barcoding)" href="20210121_reprogramming_data_merge_tags.html" />
    <link rel="prev" title="CoSpar basics" href="20210121_cospar_tutorial.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="about.html">About CoSpar</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_cospar_tutorial.html">CoSpar basics</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Hematopoietic dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_merge_tags.html">Reprogramming dataset (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_no_merge_tags.html">Reprogramming dataset (dynamic barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_lung_data.html">Lung dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_static_barcoding.html">Synthetic data (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_dynamic_barcoding.html">Synthetic data (dynamic barcoding)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Hematopoietic dataset</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar_notebooks/blob/master/20210121_all_hematopoietic_data.ipynb" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Hematopoietic-dataset">
<h1>Hematopoietic dataset<a class="headerlink" href="#Hematopoietic-dataset" title="Permalink to this headline">¶</a></h1>
<p>Hematopoiesis dataset from <em>Weinreb, C., Rodriguez-Fraticelli, A., Camargo, F. D. &amp; Klein, A. M. Science 367, (2020)</em>. This dataset has 3 time points for both the clones and the state measurements.</p>
<p>This dataset is very big. Running the whole pipeline for the first time could take 1-2 hours. Most of the time is used for generating the similarity matrix, which are saved for later usage. The actual computation of transition map takes only 10 minutues.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_version</span><span class="p">()</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">3</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span> <span class="c1"># use png to reduce file size.</span>

<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">data_path</span><span class="o">=</span><span class="s1">&#39;LARRY_data&#39;</span> <span class="c1"># A relative path to save data. If not existed before, create a new one.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">figure_path</span><span class="o">=</span><span class="s1">&#39;LARRY_figure&#39;</span> <span class="c1"># A relative path to save figures. If not existed before, create a new one.</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running cospar 0.0.2 (python 3.6.12) on 2021-01-28 16:47.
</pre></div></div>
</div>
<div class="section" id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">data_path</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">data_path</span>
<span class="n">figure_path</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">figure_path</span>
<span class="n">adata_orig</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">hematopoiesis_all</span><span class="p">(</span><span class="n">data_path</span><span class="o">=</span><span class="n">data_path</span><span class="p">,</span><span class="n">figure_path</span><span class="o">=</span><span class="n">figure_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_6_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_6_0.png" style="width: 470px; height: 274px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: []
Availabel clusters: [&#39;Neutrophil&#39;, &#39;Lymphoid&#39;, &#39;Erythroid&#39;, &#39;Eos&#39;, &#39;pDC&#39;, &#39;Monocyte&#39;, &#39;undiff&#39;, &#39;Neu_Mon&#39;, &#39;Ccr7_DC&#39;, &#39;Baso&#39;, &#39;Meg&#39;, &#39;Mast&#39;]
Availabel time points: [&#39;2&#39;, &#39;4&#39;, &#39;6&#39;]
Clonal time points: [&#39;2&#39;, &#39;4&#39;, &#39;6&#39;]
</pre></div></div>
</div>
<div class="section" id="Raw-clonal-data-analysis-(without-using-state-information)">
<h3>Raw clonal data analysis (without using state information)<a class="headerlink" href="#Raw-clonal-data-analysis-(without-using-state-information)" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">color_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;red&#39;</span><span class="p">,</span><span class="s1">&#39;blue&#39;</span><span class="p">],</span><span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_9_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_9_0.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">celltype_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mDC&#39;</span><span class="p">,</span>  <span class="s1">&#39;Mast&#39;</span><span class="p">,</span> <span class="s1">&#39;Meg&#39;</span><span class="p">,</span> <span class="s1">&#39;pDC&#39;</span><span class="p">,</span> <span class="s1">&#39;Eos&#39;</span><span class="p">,</span> <span class="s1">&#39;Ly&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery&#39;</span><span class="p">,</span> <span class="s1">&#39;Baso&#39;</span><span class="p">,</span> <span class="s1">&#39;Neu&#39;</span><span class="p">,</span> <span class="s1">&#39;Mon&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span>
                <span class="n">color_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">rename_selected_fates</span><span class="o">=</span><span class="n">celltype_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_10_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_10_0.png" style="width: 281px; height: 283px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;4&#39;</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barcode_heatmap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_11_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_11_0.png" style="width: 284px; height: 433px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># select_fate_cluster=&#39;Monocyte&#39;</span>
<span class="c1"># clonal_fate_bias,clone_id=cs.pl.clonal_fate_bias(adata_orig,select_fate_cluster,</span>
<span class="c1">#             clone_size_thresh=5,N_resampling=100,compute_new=False)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Part-1:-Infer-transition-map-using-clones-from-both-day4-and-day6">
<h2>Part 1: Infer transition map using clones from both day4 and day6<a class="headerlink" href="#Part-1:-Infer-transition-map-using-clones-from-both-day4-and-day6" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Map-inference">
<h3>Map inference<a class="headerlink" href="#Map-inference" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">t</span><span class="o">=</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1">#</span>
<span class="n">selected_clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clonal_time_points</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span>
        <span class="n">noise_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Used--time: {time.time()-t}&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------Step 1: Select time points---------
--&gt; Clonal cell fraction (day 4-6): 0.8093837722229649
--&gt; Clonal cell fraction (day 6-4): 0.8744969393621699
--&gt; Numer of cells that are clonally related -- day 4: 12110  and day 6: 25858
Valid clone number &#39;FOR&#39; post selection 3047
Cell number=37968, Clone number=3047
-------Step 2: Compute the full Similarity matrix if necessary---------
-------Step 3: Optimize the transition map recursively---------
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  2.8668580055236816
--&gt; Time elapsed:  11.12159776687622
--&gt; Time elapsed:  3.948186159133911
--&gt; Time elapsed:  13.878545045852661
Compute similarity matrix: load existing data
--&gt; Time elapsed:  2.4825279712677
--&gt; Time elapsed:  7.225665092468262
--&gt; Time elapsed:  1.8797578811645508
--&gt; Time elapsed:  9.714000940322876
Compute similarity matrix: load existing data
--&gt; Time elapsed:  1.7630691528320312
--&gt; Time elapsed:  5.514636039733887
--&gt; Time elapsed:  1.5426161289215088
--&gt; Time elapsed:  5.716917991638184
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  46.65948677062988
Phase II: time elapsed --  98.13885879516602
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  37.80651021003723
Phase II: time elapsed --  102.54157018661499
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  31.309563875198364
Phase II: time elapsed --  86.41444969177246
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Used--time: 629.5148532390594
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_15_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_15_1.png" style="width: 363px; height: 241px;" />
</div>
</div>
</div>
<div class="section" id="Save-or-load-pre-computed-data">
<h3>Save or load pre-computed data<a class="headerlink" href="#Save-or-load-pre-computed-data" title="Permalink to this headline">¶</a></h3>
<p>This can be used to save maps obtained from different parameter choices. Usually, different parameter choice will result to different ‘data_des’, a prefix to define the map. Saving an adata would print the ‘data_des’, which can be used to load the corresponding map.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">save_data</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">save_map</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">load_data</span><span class="o">=</span><span class="kc">True</span>
<span class="k">if</span> <span class="n">load_data</span><span class="p">:</span><span class="c1">### Plotting</span>
    <span class="c1"># updated Jan 27, 2021</span>
    <span class="c1">#data_des=&#39;LARRY_TwoTimeClone_t*4*6&#39;### Plotting</span>
    <span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;LARRY_OneTimeClone_t*4*6&#39;</span>
    <span class="c1">#data_des=&#39;LARRY_TwoTimeClone_t*2*4*6&#39;</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">load_saved_adata</span><span class="p">(</span><span class="n">data_des</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="Plotting">
<h3>Plotting<a class="headerlink" href="#Plotting" title="Permalink to this headline">¶</a></h3>
<p>There are some common parameters</p>
<ul class="simple">
<li><p><strong>used_map_name</strong>: <em>{‘transition_map’, ‘demultiplexed_map’, ‘OT_transition_map’, ‘HighVar_transition_map’}</em></p></li>
<li><p><strong>selected_fates</strong>: selected fate clusters for things like fate map construction, DGE analysis etc. The <em>selected_fates</em> allows nested structure, e.g., <em>selected_fates=[[‘0’,‘2’],‘1’]</em> means combined cluster ‘0-2’ and cluster ‘1’</p></li>
<li><p><strong>map_backwards</strong>: <em>True</em>, assuming that the <em>selected_fates</em> are cell states belonging to later time points (t2), and map for the initial cell states; <em>False</em>, map from early states to later states.</p></li>
<li><p><strong>plot_time_points</strong>: select states at given time points for the analysis</p></li>
</ul>
<div class="section" id="Plotting-transition-profiles-for-single-cells">
<h4>Plotting transition profiles for single cells<a class="headerlink" href="#Plotting-transition-profiles-for-single-cells" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">]</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_20_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_20_0.png" style="width: 282px; height: 883px;" />
</div>
</div>
</div>
<div class="section" id="Fate-map">
<h4>Fate map<a class="headerlink" href="#Fate-map" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_22_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_22_0.png" style="width: 287px; height: 2417px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_23_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_23_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
</div>
<div class="section" id="Fate-bias">
<h4>Fate bias<a class="headerlink" href="#Fate-bias" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_intrinsic</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_25_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_25_0.png" style="width: 528px; height: 230px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_25_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_25_1.png" style="width: 600px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_26_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_26_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_26_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_26_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
</div>
<div class="section" id="Lineage-coupling-of-the-transition-map">
<h4>Lineage coupling of the transition map<a class="headerlink" href="#Lineage-coupling-of-the-transition-map" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fate_array</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">celltype_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mDC&#39;</span><span class="p">,</span>  <span class="s1">&#39;Mast&#39;</span><span class="p">,</span> <span class="s1">&#39;Meg&#39;</span><span class="p">,</span> <span class="s1">&#39;pDC&#39;</span><span class="p">,</span> <span class="s1">&#39;Eos&#39;</span><span class="p">,</span> <span class="s1">&#39;Ly&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery&#39;</span><span class="p">,</span> <span class="s1">&#39;Baso&#39;</span><span class="p">,</span> <span class="s1">&#39;Neu&#39;</span><span class="p">,</span> <span class="s1">&#39;Mon&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">fate_array</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">rename_selected_fates</span><span class="o">=</span><span class="n">celltype_names</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_28_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_28_0.png" style="width: 366px; height: 283px;" />
</div>
</div>
</div>
<div class="section" id="Dynamic-trajectory-inference">
<h4>Dynamic trajectory inference<a class="headerlink" href="#Dynamic-trajectory-inference" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_competition_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_30_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_30_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
</div>
<div class="section" id="Differential-genes-for-two-ancestor-groups">
<h4>Differential genes for two ancestor groups<a class="headerlink" href="#Differential-genes-for-two-ancestor-groups" title="Permalink to this headline">¶</a></h4>
<p>The selected cell states can be accessed at <em>adata.obs[‘DGE_cell_group_A’]</em> and <em>adata.obs[‘DGE_cell_group_B’]</em>. A DGE analysis can be performed below. It is also easy to use other methods for DGE analysis using the selected idx.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">group_A_index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;DGE_cell_group_A&#39;</span><span class="p">]</span>
<span class="n">group_B_index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;DGE_cell_group_B&#39;</span><span class="p">]</span>
<span class="n">dge_gene_A</span><span class="p">,</span> <span class="n">dge_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">differential_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">plot_gene_N</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_33_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_33_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_33_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_33_1.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_33_2.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_33_2.png" style="width: 472px; height: 171px;" />
</div>
</div>
</div>
<div class="section" id="Gene-trend-along-the-dynamic-trajectory">
<h4>Gene trend along the dynamic trajectory<a class="headerlink" href="#Gene-trend-along-the-dynamic-trajectory" title="Permalink to this headline">¶</a></h4>
<p>The selected states, combined with the target states are stored as dynamic trajectory at adata.uns[‘dynamic_trajectory’]. We can plot gene trend along this trajectory</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The results are based on pre-computed dynamic trajectories from a previous step</span>
<span class="c1"># better to use the &#39;intraclone transition map&#39;</span>

<span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gata1&#39;</span><span class="p">,</span><span class="s1">&#39;Mpo&#39;</span><span class="p">,</span> <span class="s1">&#39;Elane&#39;</span><span class="p">,</span> <span class="s1">&#39;S100a8&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compute_new</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fig_width</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">gene_exp_percentile</span><span class="o">=</span><span class="mi">99</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">plot_raw_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: LARRY_figure/LARRY_TwoTimeClone_t*4*6_fate_trajectory_pseutoTime_gene_expression_Neutrophil_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_36_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_36_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_36_2.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_36_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
</div>
<div class="section" id="Apply-other-pseudotime-methods-on-the-selected-cell-states">
<h4>Apply other pseudotime methods on the selected cell states<a class="headerlink" href="#Apply-other-pseudotime-methods-on-the-selected-cell-states" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># adata_selected can be used by other analysis package for further analysis</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_selected</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Trying to set attribute `.uns` of view, copying.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_38_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_38_1.png" style="width: 470px; height: 274px;" />
</div>
</div>
</div>
<div class="section" id="Identify-differentially-expressed-genes">
<h4>Identify differentially expressed genes<a class="headerlink" href="#Identify-differentially-expressed-genes" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">diff_gene_A</span><span class="p">,</span><span class="n">diff_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">differential_genes_for_given_fates</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span><span class="n">gene_N</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">plot_gene_N</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_40_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_40_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_40_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_40_1.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_40_2.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_40_2.png" style="width: 472px; height: 171px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Transition-map-from-only-clonal-information">
<h3>Transition map from only clonal information<a class="headerlink" href="#Transition-map-from-only-clonal-information" title="Permalink to this headline">¶</a></h3>
<p>Assume uniform transition within the same clone</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The Weinreb method: selecting uni-potent clones first; Uniform transition probability within the same clone</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_weinreb_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;weinreb_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
This method works when there are only time points and all datasets
Used clone fraction 0.6307843780767969
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_42_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_42_1.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_42_2.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_42_2.png" style="width: 310px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The naive method: no clone pre-selection. Uniform transition probability within the same clone</span>
<span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_naive_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;naive_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_43_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_43_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_43_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_43_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Part-II:-Infer-transition-map-using-clones-from-all-time-points">
<h2>Part II: Infer transition map using clones from all time points<a class="headerlink" href="#Part-II:-Infer-transition-map-using-clones-from-all-time-points" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">t</span><span class="o">=</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1">#</span>
<span class="n">selected_clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clonal_time_points</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span>
        <span class="n">noise_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Used--time: {time.time()-t}&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------Step 1: Select time points---------
--&gt; Clonal cell fraction (day 2-4): 0.48396946564885496
--&gt; Clonal cell fraction (day 4-6): 0.8093837722229649
--&gt; Clonal cell fraction (day 4-2): 0.3728111215078198
--&gt; Clonal cell fraction (day 6-4): 0.8744969393621699
--&gt; Numer of cells that are clonally related -- day 2: 2219  and day 4: 5578
--&gt; Numer of cells that are clonally related -- day 4: 12110  and day 6: 25858
Valid clone number &#39;FOR&#39; post selection 3617
Cell number=41455, Clone number=3617
-------Step 2: Compute the full Similarity matrix if necessary---------
-------Step 3: Optimize the transition map recursively---------
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  4.178639888763428
--&gt; Time elapsed:  16.909836053848267
--&gt; Time elapsed:  7.051729202270508
--&gt; Time elapsed:  29.396594047546387
Compute similarity matrix: load existing data
--&gt; Time elapsed:  3.1903891563415527
--&gt; Time elapsed:  9.919449090957642
--&gt; Time elapsed:  3.2668142318725586
--&gt; Time elapsed:  14.067722797393799
Compute similarity matrix: load existing data
--&gt; Time elapsed:  2.4883570671081543
--&gt; Time elapsed:  7.939680099487305
--&gt; Time elapsed:  2.008693218231201
--&gt; Time elapsed:  11.060231924057007
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  76.52748799324036
Phase II: time elapsed --  149.31560707092285
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  50.76579403877258
Phase II: time elapsed --  114.15616798400879
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  40.142027139663696
Phase II: time elapsed --  105.41105198860168
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Used--time: 754.9322369098663
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_45_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_45_1.png" style="width: 363px; height: 241px;" />
</div>
</div>
<p>From our experience, it is better to use the intra-clone transition map to compute the trajectory</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># combine all time points</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_48_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_48_0.png" style="width: 500px; height: 237px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot each tiem point separately, use the intra-clone transition map.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_49_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_49_0.png" style="width: 733px; height: 237px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The results are based on pre-computed dynamic trajectories from a previous step</span>
<span class="c1"># better to use the &#39;intraclone transition map&#39;</span>

<span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gata1&#39;</span><span class="p">,</span><span class="s1">&#39;Mpo&#39;</span><span class="p">,</span> <span class="s1">&#39;Elane&#39;</span><span class="p">,</span> <span class="s1">&#39;S100a8&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Neutrophil&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compute_new</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fig_width</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">gene_exp_percentile</span><span class="o">=</span><span class="mi">99</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">plot_raw_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: LARRY_figure/LARRY_TwoTimeClone_t*2*4*6_fate_trajectory_pseutoTime_gene_expression_Neutrophil_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_51_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_51_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_51_2.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_51_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
</div>
<div class="section" id="Part-IIII:-Infer-transition-map-using-only-clones-from-the-last-time-point">
<h2>Part IIII: Infer transition map using only clones from the last time point<a class="headerlink" href="#Part-IIII:-Infer-transition-map-using-only-clones-from-the-last-time-point" title="Permalink to this headline">¶</a></h2>
<p><em>Purpose</em>: Infer transition map from scRNAseq data where cells at one time point are clonally labeled. After initializing the map by either <em>OT</em> method or <em>HighVar</em> method, We jointly infer the likely clonal ancestors and the transition map between cell states in these two time points.</p>
<p>Parameters relevant for cell state selection: <em>initial_time_points, clonal_time_point, use_full_Smatrix</em>.</p>
<p>Choose the initialization method, and set the corresponding parameters.</p>
<ul class="simple">
<li><p>‘OT’: tend to be more accurate, but not reliable under batch effect. Key parameters: <em>OT_epsilon</em>.</p></li>
<li><p>‘HighVar’: is robust to batch effect. Related parameters: <em>HighVar_gene_pctl</em>.</p></li>
</ul>
<p>Parameters relevant for CoSpar itself: <em>smooth_array, normalization_mode, CoSpar_KNN, noise_threshold, Clone_update_iter_N</em>.</p>
<p>cs.hf.check_available_choices(adata_orig)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">]</span> <span class="c1">#[&#39;2&#39;,&#39;4&#39;]</span>
<span class="n">clonal_time_point</span><span class="o">=</span><span class="s1">&#39;6&#39;</span>
<span class="n">t</span><span class="o">=</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_one_time_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">clonal_time_point</span><span class="p">,</span><span class="n">Clone_update_iter_N</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">initialize_method</span><span class="o">=</span><span class="s1">&#39;OT&#39;</span><span class="p">,</span><span class="n">OT_cost</span><span class="o">=</span><span class="s1">&#39;SPD&#39;</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;Used--time: {time.time()-t}&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 4
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Load pre-computed shortest path distance matrix
Load pre-computed custon OT matrix
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (44531, 5000)
--&gt; After excluding zero-sized clones: (44531, 3949)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
--&gt; Inferring early clonal states: current clone id 400
--&gt; Inferring early clonal states: current clone id 500
--&gt; Inferring early clonal states: current clone id 600
--&gt; Inferring early clonal states: current clone id 700
--&gt; Inferring early clonal states: current clone id 800
--&gt; Inferring early clonal states: current clone id 900
--&gt; Inferring early clonal states: current clone id 1000
--&gt; Inferring early clonal states: current clone id 1100
--&gt; Inferring early clonal states: current clone id 1200
--&gt; Inferring early clonal states: current clone id 1300
--&gt; Inferring early clonal states: current clone id 1400
--&gt; Inferring early clonal states: current clone id 1500
--&gt; Inferring early clonal states: current clone id 1600
--&gt; Inferring early clonal states: current clone id 1700
--&gt; Inferring early clonal states: current clone id 1800
--&gt; Inferring early clonal states: current clone id 1900
--&gt; Inferring early clonal states: current clone id 2000
--&gt; Inferring early clonal states: current clone id 2100
--&gt; Inferring early clonal states: current clone id 2200
--&gt; Inferring early clonal states: current clone id 2300
--&gt; Inferring early clonal states: current clone id 2400
--&gt; Inferring early clonal states: current clone id 2500
--&gt; Inferring early clonal states: current clone id 2600
--&gt; Inferring early clonal states: current clone id 2700
--&gt; Inferring early clonal states: current clone id 2800
--&gt; Inferring early clonal states: current clone id 2900
--&gt; Inferring early clonal states: current clone id 3000
--&gt; Inferring early clonal states: current clone id 3100
--&gt; Inferring early clonal states: current clone id 3200
--&gt; Inferring early clonal states: current clone id 3300
--&gt; Inferring early clonal states: current clone id 3400
--&gt; Inferring early clonal states: current clone id 3500
--&gt; Inferring early clonal states: current clone id 3600
--&gt; Inferring early clonal states: current clone id 3700
--&gt; Inferring early clonal states: current clone id 3800
--&gt; Inferring early clonal states: current clone id 3900
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  3.6797239780426025
--&gt; Time elapsed:  13.742685794830322
--&gt; Time elapsed:  4.610268831253052
--&gt; Time elapsed:  12.055868864059448
Compute similarity matrix: load existing data
--&gt; Time elapsed:  3.0670039653778076
--&gt; Time elapsed:  8.03362512588501
--&gt; Time elapsed:  3.0305347442626953
--&gt; Time elapsed:  10.189287900924683
Compute similarity matrix: load existing data
--&gt; Time elapsed:  2.1817760467529297
--&gt; Time elapsed:  6.402044057846069
--&gt; Time elapsed:  1.544910192489624
--&gt; Time elapsed:  9.86088514328003
Current iteration: 0
Use smooth_round=20
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  54.03305506706238
Phase II: time elapsed --  117.08894395828247
Current iteration: 1
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  47.91264796257019
Phase II: time elapsed --  111.59352087974548
Current iteration: 2
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Start to smooth the refined clonal map
Phase I: time elapsed --  34.98528003692627
Phase II: time elapsed --  98.14115309715271
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Clone id: 1000
--&gt; Clone id: 2000
--&gt; Clone id: 3000
Finishing computing transport map from CoSpar using inferred clonal data, used time 698.4034938812256
Used--time: 819.3050382137299
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_57_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_57_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;4&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_59_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_59_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_59_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_59_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">,</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_60_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_60_0.png" style="width: 315px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_60_1.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_60_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fate_array</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">,</span><span class="s1">&#39;Mast&#39;</span><span class="p">,</span><span class="s1">&#39;Meg&#39;</span><span class="p">,</span><span class="s1">&#39;pDC&#39;</span><span class="p">,</span><span class="s1">&#39;Eos&#39;</span><span class="p">,</span><span class="s1">&#39;Lymphoid&#39;</span><span class="p">,</span><span class="s1">&#39;Erythroid&#39;</span><span class="p">,</span><span class="s1">&#39;Baso&#39;</span><span class="p">,</span>  <span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span> <span class="s1">&#39;Monocyte&#39;</span><span class="p">]</span>
<span class="n">celltype_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mDC&#39;</span><span class="p">,</span>  <span class="s1">&#39;Mast&#39;</span><span class="p">,</span> <span class="s1">&#39;Meg&#39;</span><span class="p">,</span> <span class="s1">&#39;pDC&#39;</span><span class="p">,</span> <span class="s1">&#39;Eos&#39;</span><span class="p">,</span> <span class="s1">&#39;Ly&#39;</span><span class="p">,</span> <span class="s1">&#39;Ery&#39;</span><span class="p">,</span> <span class="s1">&#39;Baso&#39;</span><span class="p">,</span> <span class="s1">&#39;Neu&#39;</span><span class="p">,</span> <span class="s1">&#39;Mon&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_Tmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="n">fate_array</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
                        <span class="n">rename_selected_fates</span><span class="o">=</span><span class="n">celltype_names</span><span class="p">,</span><span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_62_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_62_0.png" style="width: 366px; height: 283px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_competition_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Neutrophil&#39;</span><span class="p">,</span><span class="s1">&#39;Monocyte&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_all_hematopoietic_data_64_0.png" class="no-scaled-link" src="_images/20210121_all_hematopoietic_data_64_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<p>The results are saved in adata, and can be used to perform DGE analysis or plot gene trends, as shown above.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="20210121_reprogramming_data_merge_tags.html" class="btn btn-neutral float-right" title="Reprogramming dataset (static barcoding)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="20210121_cospar_tutorial.html" class="btn btn-neutral float-left" title="CoSpar basics" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>