

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Reprogramming dataset (static barcoding) &mdash; CoSpar 0.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Reprogramming dataset (dynamic barcoding)" href="20210121_reprogramming_data_no_merge_tags.html" />
    <link rel="prev" title="Hematopoietic dataset" href="20210121_all_hematopoietic_data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="about.html">About CoSpar</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_cospar_tutorial.html">CoSpar basics</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="20210121_all_hematopoietic_data.html">Hematopoietic dataset</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Reprogramming dataset (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_no_merge_tags.html">Reprogramming dataset (dynamic barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_lung_data.html">Lung dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_static_barcoding.html">Synthetic data (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_dynamic_barcoding.html">Synthetic data (dynamic barcoding)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Reprogramming dataset (static barcoding)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar_notebooks/blob/master/20210121_reprogramming_data_merge_tags.ipynb" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Reprogramming-dataset-(static-barcoding)">
<h1>Reprogramming dataset (static barcoding)<a class="headerlink" href="#Reprogramming-dataset-(static-barcoding)" title="Permalink to this headline">¶</a></h1>
<p>The reprogramming dataset from <em>Biddy, B. A. et al. Nature 564, 219–224 (2018)</em>. This dataset has multiple time points for both the clones and the state measurements.</p>
<p>The cells are barcoded over 3 rounds during the entire differentiation process. We combine up to 3 tags from the same cell into a single clonal label in representing the X_clone matrix. In this representation, each cell has at most one clonal label. Effectively, we convert the barcodes into static labels that do not carry temporal information.</p>
<p><strong>Key components</strong>:</p>
<ul class="simple">
<li><p>Part 1: Infer transition map using clones from all time points</p></li>
<li><p>Part II: Infer transition map using clones from the last time point</p></li>
<li><p>Part III: Infer transition map from state information alone</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_version</span><span class="p">()</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span><span class="o">=</span><span class="mi">3</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">75</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span> <span class="c1"># use png to reduce file size.</span>

<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">data_path</span><span class="o">=</span><span class="s1">&#39;CellTag_data_1&#39;</span> <span class="c1"># A relative path to save data. If not existed before, create a new one.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">figure_path</span><span class="o">=</span><span class="s1">&#39;CellTag_figure_1&#39;</span> <span class="c1"># A relative path to save figures. If not existed before, create a new one.</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running cospar 0.0.2 (python 3.6.12) on 2021-01-31 19:05.
</pre></div></div>
</div>
<div class="section" id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_orig</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">reprogramming_merge_tags</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
creating directory CellTag_data_1/ for saving data
creating directory CellTag_figure_1/ for saving figures
try downloading from url
https://kleintools.hms.harvard.edu/tools/downloads/cospar/CellTagging_ConcatenateClone_adata_preprocessed.h5ad
... this may take a while but only happens once
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "40c811d9a5274dde97d95f911fe9aad3", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;state_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_6_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_6_0.png" style="width: 556px; height: 274px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: []
Availabel clusters: [&#39;3&#39;, &#39;16&#39;, &#39;19&#39;, &#39;18&#39;, &#39;12&#39;, &#39;Reprogrammed&#39;, &#39;13&#39;, &#39;0&#39;, &#39;15&#39;, &#39;22&#39;, &#39;4&#39;, &#39;8&#39;, &#39;5&#39;, &#39;20&#39;, &#39;7&#39;, &#39;10&#39;, &#39;2&#39;, &#39;Failed&#39;, &#39;11&#39;, &#39;21&#39;, &#39;9&#39;]
Availabel time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
Clonal time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">re_propressing</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">re_propressing</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_highly_variable_genes</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">)</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">remove_cell_cycle_correlated_genes</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">corr_threshold</span><span class="o">=</span><span class="mf">0.03</span><span class="p">,</span><span class="n">confirm_change</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_X_pca</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">n_pca_comp</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">get_X_emb</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">umap_min_dist</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
    <span class="c1">#cs.pp.get_state_info(adata_orig,leiden_resolution=0.5) # if this is changed, the cluster name used later will be wrong.</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;time_info&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="Raw-clonal-data-analysis-(without-using-state-information)">
<h3>Raw clonal data analysis (without using state information)<a class="headerlink" href="#Raw-clonal-data-analysis-(without-using-state-information)" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#cs.pl.clones_on_manifold(adata_orig,selected_clone_list=[1],color_list=[&#39;black&#39;,&#39;red&#39;,&#39;blue&#39;],point_size=2)</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clones_on_manifold</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clone_list</span><span class="o">=</span><span class="p">[</span><span class="mi">324</span><span class="p">,</span><span class="mi">313</span><span class="p">],</span><span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span> <span class="c1">#324,313,446,716,367</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_10_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_10_0.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_10_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_10_1.png" style="width: 247px; height: 184px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;8&#39;</span>
<span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">]</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_coupling_from_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_11_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_11_0.png" style="width: 359px; height: 283px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plot_time_point</span><span class="o">=</span><span class="s1">&#39;8&#39;</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barcode_heatmap</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">plot_time_point</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="n">selected_fates</span><span class="p">,</span> <span class="n">color_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_12_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_12_0.png" style="width: 282px; height: 433px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">clonal_fate_bias</span><span class="p">,</span><span class="n">clone_id</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonal_fate_bias</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span>
    <span class="n">clone_size_thresh</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">N_resampling</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span><span class="n">compute_new</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Current clone id: 0
Current clone id: 5
Current clone id: 10
Current clone id: 15
Current clone id: 20
Current clone id: 25
Current clone id: 30
Current clone id: 35
Current clone id: 40
Current clone id: 45
Current clone id: 50
Current clone id: 55
Current clone id: 60
Current clone id: 65
Current clone id: 70
Current clone id: 75
Current clone id: 80
Current clone id: 85
Current clone id: 90
Current clone id: 95
Current clone id: 100
Current clone id: 105
Current clone id: 110
Current clone id: 115
Current clone id: 120
Current clone id: 125
Current clone id: 130
Current clone id: 135
Current clone id: 140
Current clone id: 145
Current clone id: 150
Current clone id: 155
Current clone id: 160
Current clone id: 165
Current clone id: 170
Current clone id: 175
Current clone id: 180
Current clone id: 185
Current clone id: 190
Current clone id: 195
Current clone id: 200
Current clone id: 205
Current clone id: 210
Current clone id: 215
Current clone id: 220
Current clone id: 225
Current clone id: 230
Current clone id: 235
Current clone id: 240
Current clone id: 245
Current clone id: 250
Current clone id: 255
Current clone id: 260
Current clone id: 265
Current clone id: 270
Current clone id: 275
Current clone id: 280
Current clone id: 285
Current clone id: 290
Current clone id: 295
Current clone id: 300
Current clone id: 305
Current clone id: 310
Current clone id: 315
Current clone id: 320
Current clone id: 325
Current clone id: 330
Current clone id: 335
Current clone id: 340
Current clone id: 345
Current clone id: 350
Current clone id: 355
Current clone id: 360
Current clone id: 365
Current clone id: 370
Current clone id: 375
Current clone id: 380
Current clone id: 385
Current clone id: 390
Current clone id: 395
Current clone id: 400
Current clone id: 405
Current clone id: 410
Current clone id: 415
Current clone id: 420
Current clone id: 425
Current clone id: 430
Current clone id: 435
Current clone id: 440
Current clone id: 445
Current clone id: 450
Current clone id: 455
Current clone id: 460
Current clone id: 465
Current clone id: 470
Current clone id: 475
Current clone id: 480
Current clone id: 485
Current clone id: 490
Current clone id: 495
Current clone id: 500
Current clone id: 505
Current clone id: 510
Current clone id: 515
Current clone id: 520
Current clone id: 525
Current clone id: 530
Current clone id: 535
Current clone id: 540
Current clone id: 545
Current clone id: 550
Current clone id: 555
Current clone id: 560
Current clone id: 565
Current clone id: 570
Current clone id: 575
Current clone id: 580
Current clone id: 585
Current clone id: 590
Current clone id: 595
Current clone id: 600
Current clone id: 605
Current clone id: 610
Current clone id: 615
Current clone id: 620
Current clone id: 625
Current clone id: 630
Current clone id: 635
Current clone id: 640
Current clone id: 645
Current clone id: 650
Current clone id: 655
Current clone id: 660
Current clone id: 665
Current clone id: 670
Current clone id: 675
Current clone id: 680
Current clone id: 685
Current clone id: 690
Current clone id: 695
Current clone id: 700
Current clone id: 705
Current clone id: 710
Current clone id: 715
Current clone id: 720
Current clone id: 725
Current clone id: 730
Current clone id: 735
Current clone id: 740
Current clone id: 745
Current clone id: 750
Current clone id: 755
Current clone id: 760
Current clone id: 765
Current clone id: 770
Current clone id: 775
Current clone id: 780
Current clone id: 785
Current clone id: 790
Current clone id: 795
Current clone id: 800
Current clone id: 805
Current clone id: 810
Current clone id: 815
Current clone id: 820
Current clone id: 825
Current clone id: 830
Current clone id: 835
Current clone id: 840
Current clone id: 845
Current clone id: 850
Current clone id: 855
Current clone id: 860
Current clone id: 865
Current clone id: 870
Current clone id: 875
Current clone id: 880
Current clone id: 885
Current clone id: 890
Current clone id: 895
Current clone id: 900
Current clone id: 905
Current clone id: 910
Current clone id: 915
Current clone id: 920
Current clone id: 925
Current clone id: 930
Current clone id: 935
Current clone id: 940
Current clone id: 945
Current clone id: 950
Current clone id: 955
Current clone id: 960
Current clone id: 965
Current clone id: 970
Current clone id: 975
Current clone id: 980
Current clone id: 985
Current clone id: 990
Current clone id: 995
Current clone id: 1000
Current clone id: 1005
Current clone id: 1010
Current clone id: 1015
Current clone id: 1020
Current clone id: 1025
Current clone id: 1030
Current clone id: 1035
Current clone id: 1040
Current clone id: 1045
Current clone id: 1050
Current clone id: 1055
Current clone id: 1060
Current clone id: 1065
Current clone id: 1070
Current clone id: 1075
Current clone id: 1080
Current clone id: 1085
Current clone id: 1090
Current clone id: 1095
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_13_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_13_1.png" style="width: 283px; height: 245px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_13_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_13_2.png" style="width: 302px; height: 265px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Part-1:-Infer-transition-map-using-clones-from-all-time-points">
<h2>Part 1: Infer transition map using clones from all time points<a class="headerlink" href="#Part-1:-Infer-transition-map-using-clones-from-all-time-points" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Map-inference">
<h3>Map inference<a class="headerlink" href="#Map-inference" title="Permalink to this headline">¶</a></h3>
<p>Running it the first time takes 20 mins, 17 mins of which are used to compute the similarity matrix. When it is run again, it only takes 3 mins.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span> <span class="c1">#</span>
<span class="n">selected_clonal_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;3&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;5&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">,</span><span class="s1">&#39;8&#39;</span><span class="p">]</span>
<span class="c1">#selected_clonal_time_points=[&#39;7&#39;,&#39;8&#39;]</span>
<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_multitime_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">selected_clonal_time_points</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span>
        <span class="n">noise_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">demulti_threshold</span><span class="o">=</span><span class="n">noise_threshold</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------Step 1: Select time points---------
--&gt; Clonal cell fraction (day 3-4): 0.47126436781609193
--&gt; Clonal cell fraction (day 4-5): 0.7685492801771872
--&gt; Clonal cell fraction (day 5-6): 0.7980198019801981
--&gt; Clonal cell fraction (day 6-7): 0.805352798053528
--&gt; Clonal cell fraction (day 7-8): 0.8049918831168831
--&gt; Clonal cell fraction (day 4-3): 0.07308970099667775
--&gt; Clonal cell fraction (day 5-4): 0.6072607260726073
--&gt; Clonal cell fraction (day 6-5): 0.8746958637469586
--&gt; Clonal cell fraction (day 7-6): 0.8952922077922078
--&gt; Clonal cell fraction (day 8-7): 0.9663468859448759
--&gt; Numer of cells that are clonally related -- day 3: 41  and day 4: 66
--&gt; Numer of cells that are clonally related -- day 4: 694  and day 5: 1840
--&gt; Numer of cells that are clonally related -- day 5: 2418  and day 6: 2876
--&gt; Numer of cells that are clonally related -- day 6: 2648  and day 7: 4412
--&gt; Numer of cells that are clonally related -- day 7: 3967  and day 8: 6346
Valid clone number &#39;FOR&#39; post selection 790
Cell number=17434, Clone number=790
-------Step 2: Compute the full Similarity matrix if necessary---------
Compute similarity matrix: computing new; beta=0.1
Smooth round: 1
--&gt; Time elapsed: 0.1462101936340332
Smooth round: 2
--&gt; Time elapsed: 0.6318700313568115
Smooth round: 3
--&gt; Time elapsed: 4.058271884918213
--&gt; Orignal sparsity=0.1440132780842932, Thresholding
--&gt; Final sparsity=0.08134327961173121
similarity matrix truncated (Smooth round=3):  8.030797004699707
Smooth round: 4
--&gt; Time elapsed: 10.500530004501343
--&gt; Orignal sparsity=0.2782366721346004, Thresholding
--&gt; Final sparsity=0.1435782907327997
similarity matrix truncated (Smooth round=4):  12.915031909942627
Smooth round: 5
--&gt; Time elapsed: 15.246415138244629
--&gt; Orignal sparsity=0.37387141506207067, Thresholding
--&gt; Final sparsity=0.2072778864557119
similarity matrix truncated (Smooth round=5):  8.020248889923096
Save the matrix~~~
Smooth round: 6
--&gt; Time elapsed: 21.41595482826233
--&gt; Orignal sparsity=0.45353977091439257, Thresholding
--&gt; Final sparsity=0.2689848859325567
similarity matrix truncated (Smooth round=6):  6.984263181686401
Smooth round: 7
--&gt; Time elapsed: 34.1442928314209
--&gt; Orignal sparsity=0.5228332749578013, Thresholding
--&gt; Final sparsity=0.32826160348792477
similarity matrix truncated (Smooth round=7):  7.781483173370361
Smooth round: 8
--&gt; Time elapsed: 30.63093614578247
--&gt; Orignal sparsity=0.5844685765175182, Thresholding
--&gt; Final sparsity=0.38362760964515485
similarity matrix truncated (Smooth round=8):  7.974940061569214
Smooth round: 9
--&gt; Time elapsed: 35.917556285858154
--&gt; Orignal sparsity=0.6382832981536288, Thresholding
--&gt; Final sparsity=0.4347932758186628
similarity matrix truncated (Smooth round=9):  7.4193642139434814
Smooth round: 10
--&gt; Time elapsed: 40.89142179489136
--&gt; Orignal sparsity=0.6850691009500649, Thresholding
--&gt; Final sparsity=0.48187784020689256
similarity matrix truncated (Smooth round=10):  9.044863224029541
Save the matrix~~~
Smooth round: 11
--&gt; Time elapsed: 36.58433485031128
--&gt; Orignal sparsity=0.7256668380762321, Thresholding
--&gt; Final sparsity=0.5251253879869471
similarity matrix truncated (Smooth round=11):  7.607695817947388
Smooth round: 12
--&gt; Time elapsed: 70.08605909347534
--&gt; Orignal sparsity=0.7608566277859781, Thresholding
--&gt; Final sparsity=0.5648926975432402
similarity matrix truncated (Smooth round=12):  9.646785020828247
Smooth round: 13
--&gt; Time elapsed: 65.23817300796509
--&gt; Orignal sparsity=0.7914519214016483, Thresholding
--&gt; Final sparsity=0.6015221957973469
similarity matrix truncated (Smooth round=13):  11.143070936203003
Smooth round: 14
--&gt; Time elapsed: 70.58985090255737
--&gt; Orignal sparsity=0.8182097402031955, Thresholding
--&gt; Final sparsity=0.6353063058254474
similarity matrix truncated (Smooth round=14):  8.404071807861328
Smooth round: 15
--&gt; Time elapsed: 45.05844712257385
--&gt; Orignal sparsity=0.8416467679655742, Thresholding
--&gt; Final sparsity=0.6664864314513196
similarity matrix truncated (Smooth round=15):  7.536661148071289
Save the matrix~~~
-------Step 3: Optimize the transition map recursively---------
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  2.461054801940918
--&gt; Time elapsed:  5.206609725952148
--&gt; Time elapsed:  2.061098098754883
--&gt; Time elapsed:  5.179265737533569
Compute similarity matrix: load existing data
--&gt; Time elapsed:  1.8073961734771729
--&gt; Time elapsed:  4.124726057052612
--&gt; Time elapsed:  1.3285229206085205
--&gt; Time elapsed:  4.872633934020996
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.8039090633392334
--&gt; Time elapsed:  3.964270830154419
--&gt; Time elapsed:  0.9816751480102539
--&gt; Time elapsed:  2.7839057445526123
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Relative time point pair index: 2
--&gt; Clone id: 0
--&gt; Relative time point pair index: 3
--&gt; Clone id: 0
--&gt; Relative time point pair index: 4
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  33.08131003379822
Phase II: time elapsed --  55.68743395805359
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Relative time point pair index: 2
--&gt; Clone id: 0
--&gt; Relative time point pair index: 3
--&gt; Clone id: 0
--&gt; Relative time point pair index: 4
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  14.042507886886597
Phase II: time elapsed --  31.260063886642456
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Relative time point pair index: 2
--&gt; Clone id: 0
--&gt; Relative time point pair index: 3
--&gt; Clone id: 0
--&gt; Relative time point pair index: 4
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  12.665872812271118
Phase II: time elapsed --  29.531916856765747
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
--&gt; Relative time point pair index: 1
--&gt; Clone id: 0
--&gt; Relative time point pair index: 2
--&gt; Clone id: 0
--&gt; Relative time point pair index: 3
--&gt; Clone id: 0
--&gt; Relative time point pair index: 4
--&gt; Clone id: 0
-----------Total used time: 1220.7262461185455 s ------------
Used--time: 1220.7797009944916
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_17_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_17_1.png" style="width: 363px; height: 241px;" />
</div>
</div>
</div>
<div class="section" id="Save-or-load-pre-computed-data">
<h3>Save or load pre-computed data<a class="headerlink" href="#Save-or-load-pre-computed-data" title="Permalink to this headline">¶</a></h3>
<p>This can be used to save maps obtained from different parameter choices. Usually, different parameter choice will result to different ‘data_des’, a prefix to define the map. Saving an adata would print the ‘data_des’, which can be used to load the corresponding map.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">save_data</span><span class="o">=</span><span class="kc">True</span>
<span class="k">if</span> <span class="n">save_data</span><span class="p">:</span>
    <span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">save_map</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">load_data</span><span class="o">=</span><span class="kc">False</span>
<span class="k">if</span> <span class="n">load_data</span><span class="p">:</span>
    <span class="c1">## Updated Jan 28, 2021</span>
    <span class="c1">#data_des=&#39;CellTagging_TwoTimeClone_t*3*4*5*6*7*8&#39;</span>
    <span class="n">data_des</span><span class="o">=</span><span class="s1">&#39;CellTagging_OneTimeClone_t*3*4*5*6*7*8&#39;</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">load_saved_adata_with_key</span><span class="p">(</span><span class="n">data_des</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Saved file: data_des=&#39;CellTagging_TwoTimeClone_t*3*4*5*6*7*8&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: [&#39;transition_map&#39;, &#39;intraclone_transition_map&#39;]
Availabel clusters: [&#39;3&#39;, &#39;16&#39;, &#39;19&#39;, &#39;18&#39;, &#39;12&#39;, &#39;Reprogrammed&#39;, &#39;13&#39;, &#39;0&#39;, &#39;15&#39;, &#39;22&#39;, &#39;4&#39;, &#39;8&#39;, &#39;5&#39;, &#39;20&#39;, &#39;7&#39;, &#39;10&#39;, &#39;2&#39;, &#39;Failed&#39;, &#39;11&#39;, &#39;21&#39;, &#39;9&#39;]
Availabel time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
Clonal time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Plotting">
<h2>Plotting<a class="headerlink" href="#Plotting" title="Permalink to this headline">¶</a></h2>
<p>There are some common parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">used_map_name</span></code> (str). Determines which transition map to use for analysis. Choices: {‘transition_map’, ‘intraclone_transition_map’, ‘OT_transition_map’, ‘HighVar_transition_map’,’clonal_transition_map’}</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">selected_fates</span></code> (list of str). Selected clusters to aggregate differentiation dynamics and visualize fate bias etc.. The selected_fates allows nested structure, e.g., selected_fates=[‘1’, [‘0’, ‘2’]] selects two clusters: cluster ‘1’ and the other that combines ‘0’ and ‘2’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">map_backwards</span></code> (bool). We can analyze either the forward transitions, i.e., where the selected states or clusters are going (map_backwards=False), or the backward transitions, i.e., where these selected states or clusters came from (map_backwards=False). The latter is more useful, and is the default.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">plot_time_points</span></code> (list). List of time points to use. By default, all are used.</p></li>
</ul>
<div class="section" id="Plotting-transition-profiles-for-single-cells">
<h3>Plotting transition profiles for single cells<a class="headerlink" href="#Plotting-transition-profiles-for-single-cells" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="p">[</span><span class="mi">1000</span><span class="p">,</span><span class="mi">3500</span><span class="p">,</span><span class="mi">6000</span><span class="p">,</span><span class="mi">5500</span><span class="p">]</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">single_cell_transition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_state_id_list</span><span class="o">=</span><span class="n">selected_state_id_list</span><span class="p">,</span>
                                    <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_23_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_23_0.png" style="width: 282px; height: 883px;" />
</div>
</div>
<div class="section" id="Fate-map">
<h4>Fate map<a class="headerlink" href="#Fate-map" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_25_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_25_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
</div>
<div class="section" id="Relative-fate-bias">
<h4>Relative fate bias<a class="headerlink" href="#Relative-fate-bias" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_intrinsic</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_27_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_27_0.png" style="width: 528px; height: 230px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_27_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_27_1.png" style="width: 600px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;7&#39;</span><span class="p">],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_28_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_28_0.png" style="width: 318px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_28_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_28_1.png" style="width: 302px; height: 266px;" />
</div>
</div>
</div>
<div class="section" id="Identify-differentially-expressed-genes">
<h4>Identify differentially expressed genes<a class="headerlink" href="#Identify-differentially-expressed-genes" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_competition_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_30_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_30_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
</div>
<div class="section" id="Differential-genes-for-two-ancestor-groups">
<h4>Differential genes for two ancestor groups<a class="headerlink" href="#Differential-genes-for-two-ancestor-groups" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">dge_gene_A</span><span class="p">,</span> <span class="n">dge_gene_B</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">differential_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">plot_gene_N</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_32_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_32_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_32_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_32_1.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_32_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_32_2.png" style="width: 472px; height: 171px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># All, ranked, DGE genes for group A</span>
<span class="n">dge_gene_A</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>pv</th>
      <th>mean_1</th>
      <th>mean_2</th>
      <th>ratio</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>Apoa1</td>
      <td>1.061060e-245</td>
      <td>5.941834</td>
      <td>123.686348</td>
      <td>-4.166843</td>
    </tr>
    <tr>
      <th>29</th>
      <td>Ctgf</td>
      <td>2.061783e-173</td>
      <td>0.865551</td>
      <td>6.773607</td>
      <td>-2.058983</td>
    </tr>
    <tr>
      <th>511</th>
      <td>Kng1</td>
      <td>7.617055e-25</td>
      <td>0.899244</td>
      <td>5.939509</td>
      <td>-1.869408</td>
    </tr>
    <tr>
      <th>43</th>
      <td>S100a13</td>
      <td>3.075840e-144</td>
      <td>1.659719</td>
      <td>7.623402</td>
      <td>-1.696983</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Timp3</td>
      <td>2.261970e-284</td>
      <td>5.562981</td>
      <td>19.726454</td>
      <td>-1.659050</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>204</th>
      <td>Tpd52</td>
      <td>1.026880e-55</td>
      <td>0.486977</td>
      <td>1.223368</td>
      <td>-0.580364</td>
    </tr>
    <tr>
      <th>354</th>
      <td>Mmp15</td>
      <td>1.016207e-35</td>
      <td>0.065997</td>
      <td>0.587256</td>
      <td>-0.574332</td>
    </tr>
    <tr>
      <th>269</th>
      <td>Calml4</td>
      <td>6.892886e-45</td>
      <td>0.731211</td>
      <td>1.570095</td>
      <td>-0.570040</td>
    </tr>
    <tr>
      <th>270</th>
      <td>Cfb</td>
      <td>7.787499e-45</td>
      <td>0.647714</td>
      <td>1.445904</td>
      <td>-0.569902</td>
    </tr>
    <tr>
      <th>331</th>
      <td>Rasgef1b</td>
      <td>3.800116e-38</td>
      <td>0.188547</td>
      <td>0.762722</td>
      <td>-0.568606</td>
    </tr>
  </tbody>
</table>
<p>100 rows × 5 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Map-trajectory-backwards-iteratively">
<h4>Map trajectory backwards iteratively<a class="headerlink" href="#Map-trajectory-backwards-iteratively" title="Permalink to this headline">¶</a></h4>
<p>From our experience, it is better to use the intra-clone transition map to compute the trajectory</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot each tiem point separately, use the transition map.</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Failed&#39;</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_36_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_36_0.png" style="width: 1432px; height: 237px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot each tiem point separately, use the transition map.</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_37_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_37_0.png" style="width: 1432px; height: 237px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot each tiem point separately, use the intra-clone transition map.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span>
                        <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_38_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_38_0.png" style="width: 1432px; height: 237px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># plot each tiem point separately, use the intra-clone transition map.</span>
<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_via_iterative_mapping</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Failed&#39;</span><span class="p">,</span>
                        <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;intraclone_transition_map&#39;</span><span class="p">,</span>
    <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">map_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">plot_separately</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">apply_time_constaint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_39_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_39_0.png" style="width: 1432px; height: 237px;" />
</div>
</div>
</div>
<div class="section" id="Gene-trend-along-the-trajectory">
<h4>Gene trend along the trajectory<a class="headerlink" href="#Gene-trend-along-the-trajectory" title="Permalink to this headline">¶</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The results are based on pre-computed dynamic trajectories from the immediate previous step</span>
<span class="c1"># better to use the &#39;intraclone transition map&#39;</span>

<span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Col1a2&#39;</span><span class="p">,</span><span class="s1">&#39;Apoa1&#39;</span><span class="p">,</span><span class="s1">&#39;Peg3&#39;</span><span class="p">,</span><span class="s1">&#39;Spint2&#39;</span><span class="p">,</span><span class="s1">&#39;Mettl7a1&#39;</span><span class="p">,</span><span class="s1">&#39;Cdh1&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Failed&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compute_new</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fig_width</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">gene_exp_percentile</span><span class="o">=</span><span class="mi">99</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">plot_raw_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: CellTag_figure_1/CellTagging_TwoTimeClone_t*3*4*5*6*7*8_fate_trajectory_pseutoTime_gene_expression_Failed_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_41_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_41_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_41_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_41_2.png" style="width: 702px; height: 429px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The results are based on pre-computed dynamic trajectories from a previous step</span>
<span class="c1"># better to use the &#39;intraclone transition map&#39;</span>

<span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Col1a2&#39;</span><span class="p">,</span><span class="s1">&#39;Apoa1&#39;</span><span class="p">,</span><span class="s1">&#39;Peg3&#39;</span><span class="p">,</span><span class="s1">&#39;Spint2&#39;</span><span class="p">,</span><span class="s1">&#39;Mettl7a1&#39;</span><span class="p">,</span><span class="s1">&#39;Cdh1&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">compute_new</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fig_width</span><span class="o">=</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">gene_exp_percentile</span><span class="o">=</span><span class="mi">99</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">plot_raw_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: CellTag_figure_1/CellTagging_TwoTimeClone_t*3*4*5*6*7*8_fate_trajectory_pseutoTime_gene_expression_Reprogrammed_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_42_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_42_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_42_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_42_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Transition-map-from-only-clonal-information">
<h3>Transition map from only clonal information<a class="headerlink" href="#Transition-map-from-only-clonal-information" title="Permalink to this headline">¶</a></h3>
<p>Assume uniform transition within the same clone</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_clonal_info_alone</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;naive&#39;</span><span class="p">)</span>

<span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;clonal_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_44_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_44_0.png" style="width: 318px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_44_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_44_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
</div>
</div>
<div class="section" id="Part-II:-Infer-transition-map-using-clones-from-the-last-time-point">
<h2>Part II: Infer transition map using clones from the last time point<a class="headerlink" href="#Part-II:-Infer-transition-map-using-clones-from-the-last-time-point" title="Permalink to this headline">¶</a></h2>
<p>It takes 12 mins to compute for the first time (excluding the time for computing similarity matrix); and 5 mins later.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;3&#39;</span><span class="p">,</span><span class="s1">&#39;4&#39;</span><span class="p">,</span><span class="s1">&#39;5&#39;</span><span class="p">,</span><span class="s1">&#39;6&#39;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">]</span>
<span class="n">clonal_time_point</span><span class="o">=</span><span class="s1">&#39;8&#39;</span>

<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_one_time_clones</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">clonal_time_point</span><span class="p">,</span>
        <span class="n">initialize_method</span><span class="o">=</span><span class="s1">&#39;OT&#39;</span><span class="p">,</span><span class="n">OT_cost</span><span class="o">=</span><span class="s1">&#39;SPD&#39;</span><span class="p">,</span><span class="n">smooth_array</span><span class="o">=</span><span class="p">[</span><span class="mi">15</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span><span class="n">noise_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 3
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Compute new shortest path distance matrix
--&gt; Chosen mode is distances
Finishing computing shortest-path distance, used time 29.862331867218018
Compute new custon OT matrix
OT solver: duality_gap
Finishing computing optial transport map, used time 1.2392089366912842
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (6654, 414)
--&gt; After excluding zero-sized clones at t2: (6654, 348)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.0015919208526611328
--&gt; Time elapsed:  0.568378210067749
--&gt; Time elapsed:  0.0016720294952392578
--&gt; Time elapsed:  0.5150761604309082
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.0009791851043701172
--&gt; Time elapsed:  0.43868112564086914
--&gt; Time elapsed:  0.00109100341796875
--&gt; Time elapsed:  0.46360301971435547
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.0005650520324707031
--&gt; Time elapsed:  0.3390321731567383
--&gt; Time elapsed:  0.0009012222290039062
--&gt; Time elapsed:  0.3096041679382324
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.8483231067657471
Phase II: time elapsed --  0.850473165512085
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.5068161487579346
Phase II: time elapsed --  0.5087811946868896
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.4799518585205078
Phase II: time elapsed --  0.48163890838623047
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 37.015310287475586
-------------------------------New Start--------------------------------------------------
Current time point: 4
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Compute new shortest path distance matrix
--&gt; Chosen mode is distances
Finishing computing shortest-path distance, used time 34.604453802108765
Compute new custon OT matrix
OT solver: duality_gap
Finishing computing optial transport map, used time 11.564520120620728
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (7470, 784)
--&gt; After excluding zero-sized clones at t2: (7470, 348)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.014209985733032227
--&gt; Time elapsed:  0.46343207359313965
--&gt; Time elapsed:  0.012888669967651367
--&gt; Time elapsed:  0.42757511138916016
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.00722503662109375
--&gt; Time elapsed:  0.3724937438964844
--&gt; Time elapsed:  0.009593725204467773
--&gt; Time elapsed:  0.37938594818115234
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.005547046661376953
--&gt; Time elapsed:  0.18985390663146973
--&gt; Time elapsed:  0.007561922073364258
--&gt; Time elapsed:  0.21970820426940918
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  3.0836470127105713
Phase II: time elapsed --  3.1541261672973633
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.7445199489593506
Phase II: time elapsed --  0.8099980354309082
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.5406358242034912
Phase II: time elapsed --  0.6106979846954346
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 40.18401598930359
-------------------------------New Start--------------------------------------------------
Current time point: 5
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Compute new shortest path distance matrix
--&gt; Chosen mode is distances
Finishing computing shortest-path distance, used time 58.737828969955444
Compute new custon OT matrix
OT solver: duality_gap
Finishing computing optial transport map, used time 35.35010886192322
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (9597, 1038)
--&gt; After excluding zero-sized clones at t2: (9597, 348)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.12756109237670898
--&gt; Time elapsed:  0.43659496307373047
--&gt; Time elapsed:  0.12559223175048828
--&gt; Time elapsed:  0.4278299808502197
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.09239411354064941
--&gt; Time elapsed:  0.37059712409973145
--&gt; Time elapsed:  0.09299731254577637
--&gt; Time elapsed:  0.349837064743042
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.07923197746276855
--&gt; Time elapsed:  0.2145400047302246
--&gt; Time elapsed:  0.07623767852783203
--&gt; Time elapsed:  0.20181512832641602
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  9.086640119552612
Phase II: time elapsed --  9.66569995880127
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  1.6230919361114502
Phase II: time elapsed --  2.3093152046203613
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.8150079250335693
Phase II: time elapsed --  1.469419002532959
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 52.83109521865845
-------------------------------New Start--------------------------------------------------
Current time point: 6
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Compute new shortest path distance matrix
--&gt; Chosen mode is distances
Finishing computing shortest-path distance, used time 61.78106498718262
Compute new custon OT matrix
OT solver: duality_gap
Finishing computing optial transport map, used time 39.69208097457886
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (9855, 903)
--&gt; After excluding zero-sized clones at t2: (9855, 348)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.12160992622375488
--&gt; Time elapsed:  0.4239342212677002
--&gt; Time elapsed:  0.13516592979431152
--&gt; Time elapsed:  0.45110225677490234
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.15372896194458008
--&gt; Time elapsed:  0.34644126892089844
--&gt; Time elapsed:  0.11916685104370117
--&gt; Time elapsed:  0.34626102447509766
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.08623170852661133
--&gt; Time elapsed:  0.2085731029510498
--&gt; Time elapsed:  0.09159111976623535
--&gt; Time elapsed:  0.20116090774536133
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  10.137753009796143
Phase II: time elapsed --  10.834695100784302
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  1.6164088249206543
Phase II: time elapsed --  2.4393677711486816
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  0.8946599960327148
Phase II: time elapsed --  1.607361078262329
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 55.39346194267273
-------------------------------New Start--------------------------------------------------
Current time point: 7
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Compute new shortest path distance matrix
--&gt; Chosen mode is distances
Finishing computing shortest-path distance, used time 89.31214118003845
Compute new custon OT matrix
OT solver: duality_gap
Finishing computing optial transport map, used time 74.63937878608704
----------------
Step 2: Jointly optimize the transition map and the initial clonal states!
Joint optimization that consider possibility of clonal overlap: v2
--&gt; original clone shape: (11495, 755)
--&gt; After excluding zero-sized clones at t2: (11495, 348)
Sort clones by size (small to large)
Infer the number of initial cells to extract for each clone in advance
--&gt; Inferring early clonal states: current clone id 0
--&gt; Inferring early clonal states: current clone id 100
--&gt; Inferring early clonal states: current clone id 200
--&gt; Inferring early clonal states: current clone id 300
---------Compute the transition map-----------
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.4853849411010742
--&gt; Time elapsed:  0.8712310791015625
--&gt; Time elapsed:  0.3859226703643799
--&gt; Time elapsed:  0.5718529224395752
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.34917783737182617
--&gt; Time elapsed:  0.4595940113067627
--&gt; Time elapsed:  0.288179874420166
--&gt; Time elapsed:  0.4204981327056885
Compute similarity matrix: load existing data
--&gt; Time elapsed:  0.20076727867126465
--&gt; Time elapsed:  0.3278007507324219
--&gt; Time elapsed:  0.1556720733642578
--&gt; Time elapsed:  0.2803492546081543
Current iteration: 0
Use smooth_round=15
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  15.601884841918945
Phase II: time elapsed --  17.08461594581604
Current iteration: 1
Use smooth_round=10
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  2.6450207233428955
Phase II: time elapsed --  4.203082799911499
Current iteration: 2
Use smooth_round=5
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Start to smooth the refined clonal map
Phase I: time elapsed --  1.9045066833496094
Phase II: time elapsed --  3.522645950317383
No need for Final Smooth (i.e., clonally states are the final state space for Tmap)
----Demultiplexed transition map----
Clone normalization
--&gt; Relative time point pair index: 0
--&gt; Clone id: 0
Finishing computing transport map from CoSpar using inferred clonal data, used time 74.41073203086853
-----------Total used time: 709.7741031646729 s ------------
Total time: 709.8692650794983
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: [&#39;transition_map&#39;, &#39;OT_transition_map&#39;]
Availabel clusters: [&#39;3&#39;, &#39;16&#39;, &#39;19&#39;, &#39;18&#39;, &#39;12&#39;, &#39;Reprogrammed&#39;, &#39;13&#39;, &#39;0&#39;, &#39;15&#39;, &#39;22&#39;, &#39;4&#39;, &#39;8&#39;, &#39;5&#39;, &#39;20&#39;, &#39;7&#39;, &#39;10&#39;, &#39;2&#39;, &#39;Failed&#39;, &#39;11&#39;, &#39;21&#39;, &#39;9&#39;]
Availabel time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
Clonal time points: [&#39;7&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;6&#39;, &#39;4&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
               <span class="n">plot_color_bar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">plot_background</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">auto_color_scale</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">plot_horizontal</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_50_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_50_0.png" style="width: 538px; height: 230px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_52_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_52_0.png" style="width: 318px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_52_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_52_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_53_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_53_0.png" style="width: 318px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_53_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_53_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_competition_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_55_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_55_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<p>The selected states, combined with the target states are stored as dynamic trajectory at adata.uns[‘dynamic_trajectory’]. We can plot gene trend along this trajectory</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Col1a2&#39;</span><span class="p">,</span><span class="s1">&#39;Apoa1&#39;</span><span class="p">,</span><span class="s1">&#39;Peg3&#39;</span><span class="p">,</span><span class="s1">&#39;Spint2&#39;</span><span class="p">,</span><span class="s1">&#39;Mettl7a1&#39;</span><span class="p">,</span><span class="s1">&#39;Cdh1&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: CellTag_figure_1/CellTagging_OneTimeClone_t*3*4*5*6*7*8_fate_trajectory_pseutoTime_gene_expression_Reprogrammed_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_58_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_58_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_58_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_58_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
</div>
<div class="section" id="Part-III:-Infer-transition-map-from-state-information-alone">
<h2>Part III: Infer transition map from state information alone<a class="headerlink" href="#Part-III:-Infer-transition-map-from-state-information-alone" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">initial_time_points</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;6&#39;</span><span class="p">,</span><span class="s1">&#39;7&#39;</span><span class="p">]</span>
<span class="n">target_time_point</span><span class="o">=</span><span class="s1">&#39;8&#39;</span>

<span class="n">adata</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">tmap</span><span class="o">.</span><span class="n">infer_Tmap_from_state_info_alone</span><span class="p">(</span><span class="n">adata_orig</span><span class="p">,</span><span class="n">initial_time_points</span><span class="p">,</span><span class="n">target_time_point</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s1">&#39;OT&#39;</span><span class="p">,</span><span class="n">OT_cost</span><span class="o">=</span><span class="s1">&#39;SPD&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-------------------------------New Start--------------------------------------------------
Current time point: 6
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Load pre-computed shortest path distance matrix
Load pre-computed custon OT matrix
-------------------------------New Start--------------------------------------------------
Current time point: 7
-----------Pre-processing and sub-sampling cells------------
----------------
Step 1: Use OT method for initialization
Load pre-computed shortest path distance matrix
Load pre-computed custon OT matrix
-----------Total used time: 4.846709728240967 s ------------
Total time: 4.882646083831787
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">hf</span><span class="o">.</span><span class="n">check_available_choices</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Available transition maps: [&#39;OT_transition_map&#39;]
Availabel clusters: [&#39;7&#39;, &#39;13&#39;, &#39;10&#39;, &#39;0&#39;, &#39;Failed&#39;, &#39;8&#39;, &#39;3&#39;, &#39;5&#39;, &#39;11&#39;, &#39;21&#39;, &#39;15&#39;, &#39;16&#39;, &#39;20&#39;, &#39;12&#39;, &#39;22&#39;, &#39;9&#39;, &#39;19&#39;, &#39;18&#39;, &#39;Reprogrammed&#39;, &#39;4&#39;]
Availabel time points: [&#39;7&#39;, &#39;6&#39;, &#39;8&#39;]
Clonal time points: [&#39;7&#39;, &#39;6&#39;, &#39;8&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_map</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
               <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_63_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_63_0.png" style="width: 288px; height: 435px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">fate_bias_from_binary_competition</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span><span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span>
            <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span><span class="n">plot_target_state</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">sum_fate_prob_thresh</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_65_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_65_0.png" style="width: 318px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_65_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_65_1.png" style="width: 310px; height: 266px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dynamic_trajectory_from_intrinsic_bias</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">selected_fates</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Reprogrammed&#39;</span><span class="p">,</span><span class="s1">&#39;Failed&#39;</span><span class="p">],</span>
 <span class="n">used_map_name</span><span class="o">=</span><span class="s1">&#39;OT_transition_map&#39;</span><span class="p">,</span> <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot_time_points</span><span class="o">=</span><span class="p">[],</span>
    <span class="n">bias_threshold_A</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">bias_threshold_B</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">avoid_target_states</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_67_0.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_67_0.png" style="width: 583px; height: 283px;" />
</div>
</div>
<p>The selected states, combined with the target states are stored as dynamic trajectory at <code class="docutils literal notranslate"><span class="pre">adata.uns['dynamic_trajectory']</span></code>. We can plot gene trend along this trajectory</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gene_name_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Col1a2&#39;</span><span class="p">,</span><span class="s1">&#39;Apoa1&#39;</span><span class="p">,</span><span class="s1">&#39;Peg3&#39;</span><span class="p">,</span><span class="s1">&#39;Spint2&#39;</span><span class="p">,</span><span class="s1">&#39;Mettl7a1&#39;</span><span class="p">,</span><span class="s1">&#39;Cdh1&#39;</span><span class="p">]</span>
<span class="n">selected_fate</span><span class="o">=</span><span class="s1">&#39;Reprogrammed&#39;</span>
<span class="n">adata_selected</span><span class="o">=</span><span class="n">cs</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">gene_expression_dynamics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">selected_fate</span><span class="p">,</span><span class="n">gene_name_list</span><span class="p">,</span> <span class="n">traj_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
      <span class="n">map_backwards</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">invert_PseudoTime</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">include_target_states</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:729: PlotnineWarning: Saving 3.5 x 2.163 in image.
/Users/shouwenwang/miniconda3/envs/CoSpar_env/lib/python3.6/site-packages/plotnine/ggplot.py:730: PlotnineWarning: Filename: CellTag_figure_1/CellTagging_stateInfo_t*6*7*8_fate_trajectory_pseutoTime_gene_expression_Reprogrammed_True.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_70_1.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_70_1.png" style="width: 718px; height: 266px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/20210121_reprogramming_data_merge_tags_70_2.png" class="no-scaled-link" src="_images/20210121_reprogramming_data_merge_tags_70_2.png" style="width: 695px; height: 429px;" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="20210121_reprogramming_data_no_merge_tags.html" class="btn btn-neutral float-right" title="Reprogramming dataset (dynamic barcoding)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="20210121_all_hematopoietic_data.html" class="btn btn-neutral float-left" title="Hematopoietic dataset" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>