

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>About CoSpar &mdash; CoSpar 0.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="#" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="api.html" />
    <link rel="prev" title="CoSpar - dynamic inference by integrating transcriptome and lineage information" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">About CoSpar</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_cospar_tutorial.html">CoSpar basics</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="20210121_all_hematopoietic_data.html">All hematopoietic dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_merge_tags.html">Reprogramming dataset (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_no_merge_tags.html">Reprogramming dataset (dynamic barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_lung_data.html">Lung dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_static_barcoding.html">Synthetic data (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_dynamic_barcoding.html">Synthetic data (dynamic barcoding)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>About CoSpar</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar/blob/master/docs/source/about.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="about-cospar">
<h1>About CoSpar<a class="headerlink" href="#about-cospar" title="Permalink to this headline">¶</a></h1>
<p>High-throughput single-cell measurements have enabled unbiased study of development and differentiation, leading to numerous methods for dynamic inference. However, single-cell RNA sequencing (scRNA-seq) data alone does not fully constrain the differentiation dynamics, and existing methods inevitably operate under simplified assumptions. In parallel, the lineage information of individual cells can be profiled simultaneously along with their transcriptome by using a heritable and expressible DNA barcode as a lineage tracer. The barcode may remain static or evolve. However, sequencing is killing; we cannot get the actual single-cell dynamics in the transcriptomic space. In most cases, we can only get a snapshot of cells’ lineage information at a single time point. In some <em>in vitro</em> systems like hematopoiesis or iPS, we can approximate the dynamics by re-sampling the same clone, i.e., cells from the same ancestor, over time. The approximation is inevitably crude due to the tradeoff between the wish to observe a clone earlier and the need to allow it to expand before sampling. Therefore, clonal data also provide only partial information of the dynamics. This opens up the opportunity to integrate state and lineage (clonal) information for dynamic inference.</p>
<p>CoSpar, developed by <a class="reference external" href="https://doi.org/">Wang &amp; Klein (2021)</a>, is the first tool ever to perform dynamic inference by integrating state and lineage information. It solves for the transition probability map from cell states at an earlier time point to states at a later time point. It achieves accuracy and robustness by making use of intrinsic sparsity and coherence of the transition dynamics: neighboring initial states share similar yet sparse fate outcomes. Built upon the finite-time transition map, CoSpar can 1) infer fate potential of early states; 2) detect early fate bias among a heterogeneous progenitor population, and identify the early boundary of fate bifurcation; 3) identify putative driver genes for fate bifurcation; 4) by selecting progenitor states more accurately, we also provide a better pseudo time analysis, enabling a more accurate picture of gene expression dynamics during differentiation. CoSpar also provides several methods to analyze clonal data by itself, including the clonal coupling between fate clusters and clonal fate bias.  We envision CoSpar to be a platform to integrate key methods needed to analyze data with both state and lineage information.</p>
<div class="section" id="coherent-sparsity-optimization">
<h2>Coherent sparsity optimization<a class="headerlink" href="#coherent-sparsity-optimization" title="Permalink to this headline">¶</a></h2>
<p>One formulation of dynamic inference is to identify a finite-time transition matrix <span class="math notranslate nohighlight">\(T_{ij} (t_1,t_2)\)</span>, which describes the probability of a cell, initially in some state <span class="math notranslate nohighlight">\(i\)</span> at time <span class="math notranslate nohighlight">\(t_1\)</span>, giving rise to progeny in a state <span class="math notranslate nohighlight">\(j\)</span> at time <span class="math notranslate nohighlight">\(t_2\)</span>. However, scRNA-seq data alone do not constrain these maps fully, and strong assumptions have to be made.  We now extend these ideas to incorporate information from lineage tracing.</p>
<p>The clonal data directly constrains the transition map. We denote <em>I(t)</em> as a clone-by-cell matrix that encodes the clonal information at time <span class="math notranslate nohighlight">\(t\)</span>, and introduce  <span class="math notranslate nohighlight">\(S\)</span>  as a cell-state similarity matrix that encodes the state information. The observed clonal data is sampled from a particular realization of the stochastic differentiation dynamics. To account for this bias and technical noises, we locally smooth the raw data to obtain the “average” cell density profile per clone:  <span class="math notranslate nohighlight">\(P(t)=I(t)S\)</span>.  The transition map <span class="math notranslate nohighlight">\(T\)</span> directly links density profiles at two time points:</p>
<div class="math notranslate nohighlight">
\[\begin{equation}
P(t_2 )\approx P(t_1 )T(t_1,t_2)
\end{equation}\]</div>
<p>In most cases, the number of clones (i.e., constraints) is less than that of initial states (i.e., variables), and  <span class="math notranslate nohighlight">\(T\)</span> is not sufficiently constrained.To further constrain the map, we observe that: 1)  <span class="math notranslate nohighlight">\(T\)</span> is a sparse matrix, since most cell states have sparse differentiation outcomes; 2)  <span class="math notranslate nohighlight">\(T\)</span> is locally coherent as neighboring cell states share similar fate outcomes; 3) <span class="math notranslate nohighlight">\(T\)</span> is a non-negative matrix. With these, the inference becomes an optimization problem:</p>
<div class="math notranslate nohighlight">
\[\begin{equation}
 \min_{T} ||T||_1+\alpha ||LT||_2,  \; \text{s.t.} \; ||P(t_2)- P(t_1) T(t_1,t_2)||_{2}\le\epsilon;\; T\ge 0; \sum_j T_{ij}=1.
 \end{equation}\]</div>
<p>Here, <span class="math notranslate nohighlight">\(‖T‖_1\)</span> quantifies the sparsity of the matrix T through its l-1 norm, while  <span class="math notranslate nohighlight">\(‖LT‖_2\)</span> quantifies the local coherence of <span class="math notranslate nohighlight">\(T\)</span> (<span class="math notranslate nohighlight">\(L\)</span> is the Laplacian of the cell state similarity graph, and <span class="math notranslate nohighlight">\(LT\)</span> is the local divergence). The remaining constraints are from clonal observation, non-negativity of <span class="math notranslate nohighlight">\(T\)</span>, and map normalization, respectively. Both <span class="math notranslate nohighlight">\(\alpha\)</span> and <span class="math notranslate nohighlight">\(\epsilon\)</span> are tunable parameters.  At <span class="math notranslate nohighlight">\(\alpha=0\)</span>, the minimization takes the form of Lasso, a traditional algorithm for compressed sensing. Our formulation extends compressed sensing from vector-oriented to matrix-oriented, and improves its robustness by incorporating the local coherence constraint. The local coherence extension is reminiscent of the fused Lasso problem. We have developed CoSpar to solve the optimization.</p>
<p>The above optimization is formulated as if we have initial clonal information by re-sampling clones. When we only have the clonal information at <span class="math notranslate nohighlight">\(t_2\)</span>, we can still infer the transition map by jointly optimizing the map <span class="math notranslate nohighlight">\(T\)</span> and the initial clonal data <span class="math notranslate nohighlight">\(I(t_1)\)</span> such that the cost function is minimized. In this joint optimization, <span class="math notranslate nohighlight">\(I(t_1 )\)</span> is further constrained such that all initial cell states are labeled by clones, and non-overlapping clones at <span class="math notranslate nohighlight">\(t_2\)</span> label different cells at <span class="math notranslate nohighlight">\(t_1\)</span>.</p>
<p>See <a class="reference external" href="https://doi.org/">Wang &amp; Klein (2021)</a> for a detailed exposition of the methods.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="index.html" class="btn btn-neutral float-left" title="CoSpar - dynamic inference by integrating transcriptome and lineage information" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>