

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>API &mdash; CoSpar 0.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cospar.pp.initialize_adata_object" href="cospar.pp.initialize_adata_object.html" />
    <link rel="prev" title="About CoSpar" href="about.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About CoSpar</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_cospar_tutorial.html">CoSpar basics</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="20210121_all_hematopoietic_data.html">All hematopoietic dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_merge_tags.html">Reprogramming dataset (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_reprogramming_data_no_merge_tags.html">Reprogramming dataset (dynamic barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210121_lung_data.html">Lung dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_static_barcoding.html">Synthetic data (static barcoding)</a></li>
<li class="toctree-l1"><a class="reference internal" href="20210120-Bifurcation_model_dynamic_barcoding.html">Synthetic data (dynamic barcoding)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <a href="https://github.com/theislab/cospar/blob/master/docs/source/api.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<span class="target" id="module-cospar"></span><p>CoSpar - dynamic inference by integrating transcriptome and lineage information</p>
<div class="section" id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h1>
<p>Import CoSpar as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cospar</span> <span class="k">as</span> <span class="nn">cs</span>
</pre></div>
</div>
<p>CoSpar is built around the <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev48+g5f3be07)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object (usually called <cite>adata</cite>). For each cell, we store its RNA count matrix at <code class="docutils literal notranslate"><span class="pre">adata.X</span></code>, the gene names at <code class="docutils literal notranslate"><span class="pre">adata.var_names</span></code>,time information at <code class="docutils literal notranslate"><span class="pre">adata.obs['time_info']</span></code>, state annotation at <code class="docutils literal notranslate"><span class="pre">adata.obs['state_info']</span></code>,  clonal information at <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_clone']</span></code>, and 2-d embedding at <code class="docutils literal notranslate"><span class="pre">adata.obsm['X_emb']</span></code>.</p>
<p>Once the <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev48+g5f3be07)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object is initialized via <code class="xref py py-func docutils literal notranslate"><span class="pre">cs.pp.initialize_adata_object()</span></code>, the typical flow of analysis is to 1) perform preprocessing and dimension reduction (<code class="docutils literal notranslate"><span class="pre">cs.pp.*</span></code>); 2) visualize and analyzing clonal data alone (<code class="docutils literal notranslate"><span class="pre">cs.pl.*</span></code>); 3) infer transition map (<code class="docutils literal notranslate"><span class="pre">cs.tmap.*</span></code>); and 4) analyze inferred map using the plotting functions (<code class="docutils literal notranslate"><span class="pre">cs.pl.*</span></code>). We also provide 7 built-in datasets (<code class="docutils literal notranslate"><span class="pre">cs.datasets.*</span></code>) and miscellaneous functions to assist with the analysis (<code class="docutils literal notranslate"><span class="pre">cs.hf.*</span></code>). See <a class="reference internal" href="getting_started.html"><span class="doc">tutorial</span></a> for details.</p>
<div class="section" id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pp.initialize_adata_object.html#cospar.pp.initialize_adata_object" title="cospar.pp.initialize_adata_object"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.initialize_adata_object</span></code></a>(…[, X_clone, …])</p></td>
<td><p>Initialized the <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev48+g5f3be07)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pp.get_X_pca.html#cospar.pp.get_X_pca" title="cospar.pp.get_X_pca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.get_X_pca</span></code></a>(adata[, n_pca_comp])</p></td>
<td><p>Get X_pca.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pp.get_X_emb.html#cospar.pp.get_X_emb" title="cospar.pp.get_X_emb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.get_X_emb</span></code></a>(adata[, n_neighbors, umap_min_dist])</p></td>
<td><p>Get X_emb using <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/api/scanpy.tl.umap.html#scanpy.tl.umap" title="(in Scanpy v1.7.0rc2.dev14+g1baac547)"><code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.umap()</span></code></a></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pp.get_state_info.html#cospar.pp.get_state_info" title="cospar.pp.get_state_info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.get_state_info</span></code></a>(adata[, leiden_resolution])</p></td>
<td><p>Update <cite>state_info</cite> using <a class="reference external" href="https://scanpy.readthedocs.io/en/latest/api/scanpy.tl.leiden.html#scanpy.tl.leiden" title="(in Scanpy v1.7.0rc2.dev14+g1baac547)"><code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.tl.leiden()</span></code></a></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pp.refine_state_info_by_marker_genes.html#cospar.pp.refine_state_info_by_marker_genes" title="cospar.pp.refine_state_info_by_marker_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.refine_state_info_by_marker_genes</span></code></a>(adata, …)</p></td>
<td><p>Refine state info according to marker gene expression.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pp.refine_state_info_by_leiden_clustering.html#cospar.pp.refine_state_info_by_leiden_clustering" title="cospar.pp.refine_state_info_by_leiden_clustering"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pp.refine_state_info_by_leiden_clustering</span></code></a>(adata)</p></td>
<td><p>Refine state info by clustering states at given time points.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="transition-map-inference">
<h2>Transition map inference<a class="headerlink" href="#transition-map-inference" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.tmap.infer_Tmap_from_multitime_clones.html#cospar.tmap.infer_Tmap_from_multitime_clones" title="cospar.tmap.infer_Tmap_from_multitime_clones"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmap.infer_Tmap_from_multitime_clones</span></code></a>(…[, …])</p></td>
<td><p>Infer Tmap for clonal data with multiple time points.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.tmap.infer_intraclone_Tmap.html#cospar.tmap.infer_intraclone_Tmap" title="cospar.tmap.infer_intraclone_Tmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmap.infer_intraclone_Tmap</span></code></a>(adata[, …])</p></td>
<td><p>Infer intra-clone transition map.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.tmap.infer_Tmap_from_one_time_clones.html#cospar.tmap.infer_Tmap_from_one_time_clones" title="cospar.tmap.infer_Tmap_from_one_time_clones"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmap.infer_Tmap_from_one_time_clones</span></code></a>(…[, …])</p></td>
<td><p>Infer transition map from clones with a single time point</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.tmap.infer_Tmap_from_state_info_alone.html#cospar.tmap.infer_Tmap_from_state_info_alone" title="cospar.tmap.infer_Tmap_from_state_info_alone"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmap.infer_Tmap_from_state_info_alone</span></code></a>(…[, …])</p></td>
<td><p>Infer transition map from state information alone.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.tmap.infer_Tmap_from_clonal_info_alone.html#cospar.tmap.infer_Tmap_from_clonal_info_alone" title="cospar.tmap.infer_Tmap_from_clonal_info_alone"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmap.infer_Tmap_from_clonal_info_alone</span></code></a>(adata)</p></td>
<td><p>Compute transition map using only the lineage information</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="plotting">
<h2>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">¶</a></h2>
<p><strong>Clone analysis</strong> (clone visualization, clustering etc.)</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.clones_on_manifold.html#cospar.pl.clones_on_manifold" title="cospar.pl.clones_on_manifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.clones_on_manifold</span></code></a>(adata[, …])</p></td>
<td><p>Plot clones on top of state embedding.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.barcode_heatmap.html#cospar.pl.barcode_heatmap" title="cospar.pl.barcode_heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.barcode_heatmap</span></code></a>(adata, plot_time_point[, …])</p></td>
<td><p>Plot barcode heatmap among different fate clusters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.fate_coupling_from_clones.html#cospar.pl.fate_coupling_from_clones" title="cospar.pl.fate_coupling_from_clones"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.fate_coupling_from_clones</span></code></a>(adata, …[, …])</p></td>
<td><p>Plot fate coupling determined by just clones.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.clonal_fate_bias.html#cospar.pl.clonal_fate_bias" title="cospar.pl.clonal_fate_bias"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.clonal_fate_bias</span></code></a>(adata[, selected_fate, …])</p></td>
<td><p>Plot clonal fate bias towards a cluster.</p></td>
</tr>
</tbody>
</table>
<p><strong>Transition map analysis</strong> (fate bias etc.)</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.fate_map.html#cospar.pl.fate_map" title="cospar.pl.fate_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.fate_map</span></code></a>(adata[, selected_fates, …])</p></td>
<td><p>Plot transition probability to given fate/ancestor clusters</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.single_cell_transition.html#cospar.pl.single_cell_transition" title="cospar.pl.single_cell_transition"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.single_cell_transition</span></code></a>(adata, …[, …])</p></td>
<td><p>Plot transition probability from given initial cell states</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.fate_bias_intrinsic.html#cospar.pl.fate_bias_intrinsic" title="cospar.pl.fate_bias_intrinsic"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.fate_bias_intrinsic</span></code></a>(adata[, …])</p></td>
<td><p>Plot the relative fate bias (predicted/expected) towards a cluster.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.fate_bias_from_binary_competition.html#cospar.pl.fate_bias_from_binary_competition" title="cospar.pl.fate_bias_from_binary_competition"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.fate_bias_from_binary_competition</span></code></a>(adata)</p></td>
<td><p>Plot fate bias to given two fate/ancestor clusters (A, B).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.dynamic_trajectory_from_intrinsic_bias.html#cospar.pl.dynamic_trajectory_from_intrinsic_bias" title="cospar.pl.dynamic_trajectory_from_intrinsic_bias"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.dynamic_trajectory_from_intrinsic_bias</span></code></a>(adata)</p></td>
<td><p>Identify trajectory towards/from two given clusters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.dynamic_trajectory_from_competition_bias.html#cospar.pl.dynamic_trajectory_from_competition_bias" title="cospar.pl.dynamic_trajectory_from_competition_bias"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.dynamic_trajectory_from_competition_bias</span></code></a>(adata)</p></td>
<td><p>Identify trajectory towards/from two given clusters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.dynamic_trajectory_via_iterative_mapping.html#cospar.pl.dynamic_trajectory_via_iterative_mapping" title="cospar.pl.dynamic_trajectory_via_iterative_mapping"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.dynamic_trajectory_via_iterative_mapping</span></code></a>(…)</p></td>
<td><p>Infer trajectory towards/from a cluster</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.gene_expression_dynamics.html#cospar.pl.gene_expression_dynamics" title="cospar.pl.gene_expression_dynamics"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.gene_expression_dynamics</span></code></a>(adata, …[, …])</p></td>
<td><p>Plot gene trend along inferred dynamic trajectory</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.fate_coupling_from_Tmap.html#cospar.pl.fate_coupling_from_Tmap" title="cospar.pl.fate_coupling_from_Tmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.fate_coupling_from_Tmap</span></code></a>(adata[, …])</p></td>
<td><p>Plot fate coupling determined by Tmap.</p></td>
</tr>
</tbody>
</table>
<p><strong>Differential gene expression analysis</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.differential_genes.html#cospar.pl.differential_genes" title="cospar.pl.differential_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.differential_genes</span></code></a>(adata[, plot_groups, …])</p></td>
<td><p>Perform differential gene expression analysis and plot top DGE genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.differential_genes_for_given_fates.html#cospar.pl.differential_genes_for_given_fates" title="cospar.pl.differential_genes_for_given_fates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.differential_genes_for_given_fates</span></code></a>(adata)</p></td>
<td><p>Find and plot DGE genes between different clusters</p></td>
</tr>
</tbody>
</table>
<p><strong>General plotting</strong></p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.pl.embedding.html#cospar.pl.embedding" title="cospar.pl.embedding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.embedding</span></code></a>(adata[, basis, color])</p></td>
<td><p>Scatter plot for user specified embedding basis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.pl.gene_expression_on_manifold.html#cospar.pl.gene_expression_on_manifold" title="cospar.pl.gene_expression_on_manifold"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pl.gene_expression_on_manifold</span></code></a>(adata, …[, …])</p></td>
<td><p>Plot gene expression on the state manifold</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="datasets">
<h2>Datasets<a class="headerlink" href="#datasets" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.datasets.hematopoiesis_subsampled.html#cospar.datasets.hematopoiesis_subsampled" title="cospar.datasets.hematopoiesis_subsampled"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.hematopoiesis_subsampled</span></code></a>([data_des])</p></td>
<td><p>Top 15% most heterogeneous clones of the hematopoiesis data set from</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.datasets.hematopoiesis_all.html#cospar.datasets.hematopoiesis_all" title="cospar.datasets.hematopoiesis_all"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.hematopoiesis_all</span></code></a>([data_des])</p></td>
<td><p>All of the hematopoiesis data set from</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.datasets.lung.html#cospar.datasets.lung" title="cospar.datasets.lung"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.lung</span></code></a>([data_des])</p></td>
<td><p>The direct lung differentiation dataset from</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.datasets.reprogramming_no_merge_tags.html#cospar.datasets.reprogramming_no_merge_tags" title="cospar.datasets.reprogramming_no_merge_tags"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.reprogramming_no_merge_tags</span></code></a>([data_des])</p></td>
<td><p>The reprogramming dataset from</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.datasets.reprogramming_merge_tags.html#cospar.datasets.reprogramming_merge_tags" title="cospar.datasets.reprogramming_merge_tags"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.reprogramming_merge_tags</span></code></a>([data_des])</p></td>
<td><p>The reprogramming dataset from</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.datasets.synthetic_bifurcation_static_BC.html#cospar.datasets.synthetic_bifurcation_static_BC" title="cospar.datasets.synthetic_bifurcation_static_BC"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.synthetic_bifurcation_static_BC</span></code></a>([…])</p></td>
<td><p>Synthetic clonal datasets with static barcoding.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.datasets.synthetic_bifurcation_dynamic_BC.html#cospar.datasets.synthetic_bifurcation_dynamic_BC" title="cospar.datasets.synthetic_bifurcation_dynamic_BC"><code class="xref py py-obj docutils literal notranslate"><span class="pre">datasets.synthetic_bifurcation_dynamic_BC</span></code></a>([…])</p></td>
<td><p>Synthetic clonal datasets with dynamic barcoding.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="help-functions">
<h2>Help functions<a class="headerlink" href="#help-functions" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.hf.read.html#cospar.hf.read" title="cospar.hf.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.read</span></code></a>(filename[, backed, sheet, ext, …])</p></td>
<td><p>Read file and return <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev48+g5f3be07)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.hf.check_available_map.html#cospar.hf.check_available_map" title="cospar.hf.check_available_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.check_available_map</span></code></a>(adata)</p></td>
<td><p>Check available transition map.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.hf.save_map.html#cospar.hf.save_map" title="cospar.hf.save_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.save_map</span></code></a>(adata)</p></td>
<td><p>Save the adata and print file name prefix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.hf.save_preprocessed_adata.html#cospar.hf.save_preprocessed_adata" title="cospar.hf.save_preprocessed_adata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.save_preprocessed_adata</span></code></a>(adata[, data_des])</p></td>
<td><p>Save preprocessed adata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="cospar.hf.load_saved_adata_with_key.html#cospar.hf.load_saved_adata_with_key" title="cospar.hf.load_saved_adata_with_key"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.load_saved_adata_with_key</span></code></a>(data_des)</p></td>
<td><p>Load pre-saved adata based on the prefix ‘data_des’</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="cospar.hf.check_adata_structure.html#cospar.hf.check_adata_structure" title="cospar.hf.check_adata_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">hf.check_adata_structure</span></code></a>(adata)</p></td>
<td><p>Check whether the adata has the right structure.</p></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="cospar.pp.initialize_adata_object.html" class="btn btn-neutral float-right" title="cospar.pp.initialize_adata_object" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="about.html" class="btn btn-neutral float-left" title="About CoSpar" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>