

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>cospar.pl.driver_genes_from_intrinsic_bias &mdash; CoSpar 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cospar.pl.driver_genes_from_competition_bias" href="cospar.pl.driver_genes_from_competition_bias.html" />
    <link rel="prev" title="cospar.pl.gene_expression_on_manifold" href="cospar.pl.gene_expression_on_manifold.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About scVelo</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="20201121-Bifurcation_model_staticBC-CoSapr_basic.html">CoSpar basics with synthetic data</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="api.html">API</a> &raquo;</li>
        
      <li>cospar.pl.driver_genes_from_intrinsic_bias</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/ShouWenWang/cospar/tree/master/cospar/plotting/_plotting.py#L813-L970" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="cospar-pl-driver-genes-from-intrinsic-bias">
<h1>cospar.pl.driver_genes_from_intrinsic_bias<a class="headerlink" href="#cospar-pl-driver-genes-from-intrinsic-bias" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="cospar.pl.driver_genes_from_intrinsic_bias">
<code class="sig-prename descclassname">cospar.pl.</code><code class="sig-name descname">driver_genes_from_intrinsic_bias</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">selected_fates</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">used_map_name</span><span class="o">=</span><span class="default_value">'transition_map'</span></em>, <em class="sig-param"><span class="n">map_backwards</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">plot_time_points</span><span class="o">=</span><span class="default_value">[]</span></em>, <em class="sig-param"><span class="n">bias_threshold_A</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">bias_threshold_B</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">plot_groups</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">gene_N</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">plot_gene_N</span><span class="o">=</span><span class="default_value">5</span></em>, <em class="sig-param"><span class="n">savefig</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">point_size</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">avoid_target_states</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cospar.pl.driver_genes_from_intrinsic_bias" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot genes that correlate with fate bias towards/from given clusters.</p>
<p>If <cite>map_backwards=True</cite>, use the fate probability towards given clusters (A,B)
to define the ancestor population for A and B. And perform Differential Gene
Expression analysis (DGE) to find leading differentially expressed genes between the
ancestor population. Figures are saved at defined directory at adata.uns[‘figure_path’].</p>
<p>Fate bias at each state is a <cite>vector</cite>: (<cite>relative_bias_A,relative_bias_B</cite>).
For a given cluster, is defined by comparing predicted fate probability with
expected one, same as <a class="reference internal" href="cospar.pl.fate_map.html#cospar.pl.fate_map" title="cospar.pl.fate_map"><code class="xref py py-func docutils literal notranslate"><span class="pre">fate_map()</span></code></a>. It sales from [0,1], with 0.5 as the
neutral prediction that agrees with null hypothesis. Selected ancestors satisfy:</p>
<blockquote>
<div><ul class="simple">
<li><p>For A: <cite>relative_bias_A</cite> &gt; <cite>bias_threshold_A</cite></p></li>
<li><p>For B: <cite>relative_bias_B</cite> &gt; <cite>bias_threshold_B</cite></p></li>
</ul>
</div></blockquote>
<p>If <cite>map_backwards=False</cite>, use the probability from given clusters (A,B)
to define the fate population, and perform DGE analysis. This is not useful.</p>
<p><cite>selected_fates</cite> could contain a nested list of clusters. If so, we
combine each sub list into a mega-fate cluster and combine the fate
map correspondingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>adata</strong> : <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev44+gf1d1e17)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object</dt><dd><p>Assume to contain transition maps at adata.uns.</p>
</dd>
<dt><strong>selected_fates</strong> : <cite>list</cite></dt><dd><p>List of cluster_ids consistent with adata.obs[‘state_info’]</p>
</dd>
<dt><strong>used_map_name</strong> : <cite>str</cite></dt><dd><p>The transition map to be used for plotting: {‘transition_map’,
‘intraclone_transition_map’,’weinreb_transition_map’,’naive_transition_map’,
‘OT_transition_map’,’HighVar_transition_map’}. The actual available
map depends on adata itself, which can be accessed at adata.uns[‘available_map’]</p>
</dd>
<dt><strong>map_backwards</strong> : <cite>bool</cite>, optional (default: True)</dt><dd><p>If <cite>map_backwards=True</cite>, plot initial cell states (rows of Tmap, at t1);
else, plot later cell states (columns of Tmap, at t2)</p>
</dd>
<dt><strong>plot_time_points</strong> : <cite>list</cite>, optional (default: all)</dt><dd><p>A list of time points to further restrict the cell states to plot.
If map_backwards=True, plot initial states that are among these time points;
else, plot later states that are among these time points.</p>
</dd>
<dt><strong>bias_threshold_A</strong> : <cite>float</cite>, optional (default: 0.5)</dt><dd><p>Relative fate bias threshold to select ancestor population for cluster A.
The 0.5 is the neural bias point. Range: [0,1].
Closer to 1 means more stringent selection.</p>
</dd>
<dt><strong>bias_threshold_B</strong> : <cite>float</cite>, optional (default: 0.5)</dt><dd><p>Relative fate bias threshold to select ancestor population for cluster B.
Closer to 1 means more stringent selection.</p>
</dd>
<dt><strong>plot_groups</strong> : <cite>bool</cite>, optional (default: True)</dt><dd><p>If true, plot the selected ancestor states for A, B</p>
</dd>
<dt><strong>gene_N</strong> : <cite>int</cite>, optional (default: 100)</dt><dd><p>Number of top differentially expressed genes to selected.</p>
</dd>
<dt><strong>plot_gene_N</strong> : <cite>int</cite>, optional (default: 5)</dt><dd><p>Number of top DGE genes to plot</p>
</dd>
<dt><strong>savefig</strong> : <cite>bool</cite>, optional (default: False)</dt><dd><p>Save all plots.</p>
</dd>
<dt><strong>point_size</strong> : <cite>int</cite>, optional (default: 2)</dt><dd><p>Size of the data point.</p>
</dd>
<dt><strong>avoid_target_states</strong> : <cite>bool</cite>, optional (default: False)</dt><dd><p>If true, avoid selecting cells at the target cluster (A, or B).</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>diff_gene_A</strong> (<cite>pd.DataFrame</cite>) – Genes differentially expressed in cell state group A, ranked
by the ratio of mean expressions between
the two groups, with the top being more differentially expressed.</p></li>
<li><p><strong>diff_gene_B</strong> (<cite>pd.DataFrame</cite>) – Genes differentially expressed in cell state group B, ranked
by the ratio of mean expressions between
the two groups, with the top being more differentially expressed.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="cospar.pl.driver_genes_from_competition_bias.html" class="btn btn-neutral float-right" title="cospar.pl.driver_genes_from_competition_bias" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="cospar.pl.gene_expression_on_manifold.html" class="btn btn-neutral float-left" title="cospar.pl.gene_expression_on_manifold" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>