

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>cospar.tmap.Tmap_from_highly_variable_genes &mdash; CoSpar 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cospar.tmap.infer_initial_states_of_a_clone" href="cospar.tmap.infer_initial_states_of_a_clone.html" />
    <link rel="prev" title="cospar.tmap.compute_custom_OT_transition_map" href="cospar.tmap.compute_custom_OT_transition_map.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> CoSpar
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="about.html">About scVelo</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="20201121-Bifurcation_model_staticBC-CoSapr_basic.html">CoSpar basics with synthetic data</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">CoSpar</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          



















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="api.html">API</a> &raquo;</li>
        
      <li>cospar.tmap.Tmap_from_highly_variable_genes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/ShouWenWang/cospar/tree/master/cospar/tmap/map_reconstruction.py#L1000-L1153" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="cospar-tmap-tmap-from-highly-variable-genes">
<h1>cospar.tmap.Tmap_from_highly_variable_genes<a class="headerlink" href="#cospar-tmap-tmap-from-highly-variable-genes" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="cospar.tmap.Tmap_from_highly_variable_genes">
<code class="sig-prename descclassname">cospar.tmap.</code><code class="sig-name descname">Tmap_from_highly_variable_genes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="n">min_counts</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">min_cells</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">min_gene_vscore_pctl</span><span class="o">=</span><span class="default_value">85</span></em>, <em class="sig-param"><span class="n">smooth_array</span><span class="o">=</span><span class="default_value">[15, 10, 5]</span></em>, <em class="sig-param"><span class="n">neighbor_N</span><span class="o">=</span><span class="default_value">20</span></em>, <em class="sig-param"><span class="n">noise_threshold</span><span class="o">=</span><span class="default_value">0.2</span></em>, <em class="sig-param"><span class="n">normalization_mode</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">use_full_Smatrix</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">trunca_threshold</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">compute_new_Smatrix</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">save_subset</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cospar.tmap.Tmap_from_highly_variable_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate Tmap based on state info using HighVar.</p>
<p>We convert differentially expressed genes into <cite>pseudo-clones</cite>,
and run cospar to infer the transition map.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>adata</strong> : <a class="reference external" href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6.dev44+gf1d1e17)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object</dt><dd><p>assumed to be preprocessed, only has two time points.</p>
</dd>
<dt><strong>min_counts</strong> : <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em><em> (</em><em>default: 3</em><em>)</em></dt><dd><p>Minimum number of UMIs per cell to be considered for selecting highly variable genes.</p>
</dd>
<dt><strong>min_cells</strong> : <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em><em> (</em><em>default: 3</em><em>)</em></dt><dd><p>Minimum number of cells per gene to be considered for selecting highly variable genes.</p>
</dd>
<dt><strong>min_gene_vscore_pctl</strong> : <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a><em>, </em><em>optional</em><em> (</em><em>default: 85</em><em>)</em></dt><dd><p>Genes wht a variability percentile higher than this threshold are marked as
highly variable genes for dimension reduction. Range: [0,100]</p>
</dd>
<dt><strong>smooth_array</strong> : <cite>list</cite>, optional (default: [15,10,5])</dt><dd><p>List of smooth rounds at each iteration.
The n-th entry determines the smooth round for the Smatrix
at the n-th iteration. Its length determins the number of
iteration.</p>
</dd>
<dt><strong>neighbor_N</strong> : <cite>int</cite>, optional (default: 20)</dt><dd><p>the number of neighbors for KNN graph used for computing the similarity matrix.</p>
</dd>
<dt><strong>trunca_threshold</strong> : <cite>float</cite>, optional (default: 0.001)</dt><dd><p>We set entries to zero in the computed similarity matrix that
are smaller than this threshold. This is to promote the Smatrix sparsity, which
leads to faster computation, and smaller file size.
This threshld should be small, but not too small.</p>
</dd>
<dt><strong>noise_threshold</strong> : <cite>float</cite>, optional (default: 0.1)</dt><dd><p>noise threshold to remove noises in the updated transition map,
in the range [0,1]</p>
</dd>
<dt><strong>normalization_mode</strong> : <cite>int</cite>, optional (default: 2)</dt><dd><p>Method for normalization. Choice: [0,1,2]
0, single-cell normalization
1, Clone normalization: N2/N1 (this one does not make sense)
2, Clone normalization</p>
</dd>
<dt><strong>save_subset</strong> : <cite>bool</cite>, optional (default: True)</dt><dd><p>If true, save only Smatrix at smooth round [5,10,15,…].
Else, save Smatrix at each round.</p>
</dd>
<dt><strong>use_full_Smatrix</strong> : <cite>bool</cite>, optional (default: False)</dt><dd><p>use the Smatrix as defined by all cells, whether they are clonally
barcoded or not. We sub-sample cell states relevant for downstream
analysis from this full Smatrix. This may refine the Smatrix.
But will also increase the computation time significantly.</p>
</dd>
<dt><strong>compute_new_Smatrix</strong> : <cite>bool</cite>, optional (default: False)</dt><dd><p>If True, compute Smatrix from scratch, whether it was
computed and saved before or not.</p>
</dd>
<dt><strong>verbose</strong> : <cite>bool</cite>, optional (default: True)</dt><dd><p>True, print information of the analysis; False, do not print.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><em>None. Results are stored at adata.uns[‘HighVar_transition_map’].</em></p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="cospar.tmap.infer_initial_states_of_a_clone.html" class="btn btn-neutral float-right" title="cospar.tmap.infer_initial_states_of_a_clone" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="cospar.tmap.compute_custom_OT_transition_map.html" class="btn btn-neutral float-left" title="cospar.tmap.compute_custom_OT_transition_map" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Shou-Wen Wang.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>